// pki-lite/dist/ocsp/CertStatus.js
import { asn1js, PkiBase } from '../core/PkiBase.js';
import { Asn1ParseError } from '../core/errors/Asn1ParseError.js';
/**
 * CertStatus ::= CHOICE {
 *     good                [0] IMPLICIT NULL,
 *     revoked             [1] IMPLICIT RevokedInfo,
 *     unknown             [2] IMPLICIT UnknownInfo
 * }
 *
 * UnknownInfo ::= NULL
 */
export class CertStatus extends PkiBase {
    static good = new CertStatus({ status: 'good' });
    static revoked = new CertStatus({ status: 'revoked' });
    static unknown = new CertStatus({ status: 'unknown' });
    status;
    revocationTime;
    revocationReason;
    constructor(options) {
        super();
        this.status = options.status;
        this.revocationTime = options.revocationTime;
        this.revocationReason = options.revocationReason;
    }
    static createGood() {
        return new CertStatus({
            status: 'good',
        });
    }
    static createRevoked(revocationTime, revocationReason) {
        return new CertStatus({
            status: 'revoked',
            revocationTime,
            revocationReason,
        });
    }
    static createUnknown() {
        return new CertStatus({
            status: 'unknown',
        });
    }
    toAsn1() {
        if (this.status === 'good') {
            return new asn1js.Primitive({
                idBlock: {
                    tagClass: 3, // CONTEXT-SPECIFIC
                    tagNumber: 0, // [0]
                },
                lenBlock: { length: 1 },
            });
        }
        else if (this.status === 'revoked') {
            const values = [
                new asn1js.GeneralizedTime({
                    valueDate: this.revocationTime || new Date(),
                }),
            ];
            if (this.revocationReason !== undefined) {
                values.push(new asn1js.Constructed({
                    idBlock: {
                        tagClass: 3, // CONTEXT-SPECIFIC
                        tagNumber: 0, // [0]
                    },
                    value: [
                        new asn1js.Enumerated({
                            value: this.revocationReason,
                        }),
                    ],
                }));
            }
            const revokedInfo = new asn1js.Sequence({
                value: values,
            });
            return new asn1js.Constructed({
                idBlock: {
                    tagClass: 3, // CONTEXT-SPECIFIC
                    tagNumber: 1, // [1]
                },
                value: [revokedInfo],
            });
        }
        else {
            // unknown
            return new asn1js.Primitive({
                idBlock: {
                    tagClass: 3, // CONTEXT-SPECIFIC
                    tagNumber: 2, // [2]
                },
                lenBlock: { length: 1 },
            });
        }
    }
    static fromAsn1(asn1) {
        if (asn1 instanceof asn1js.Primitive &&
            asn1.idBlock.tagClass === 3 &&
            asn1.idBlock.tagNumber === 0) {
            // good
            return CertStatus.createGood();
        }
        else if (asn1 instanceof asn1js.Constructed &&
            asn1.idBlock.tagClass === 3 &&
            asn1.idBlock.tagNumber === 1) {
            // revoked
            const revokedInfo = asn1.valueBlock.value[0];
            const revocationTimeBlock = revokedInfo.valueBlock
                .value[0];
            // Parse the date from the GeneralizedTime
            const revocationTime = revocationTimeBlock.toDate();
            let revocationReason = undefined;
            if (revokedInfo.valueBlock.value.length > 1) {
                const reasonContainer = revokedInfo.valueBlock
                    .value[1];
                if (reasonContainer.idBlock.tagClass === 3 &&
                    reasonContainer.idBlock.tagNumber === 0) {
                    const reasonValue = reasonContainer.valueBlock
                        .value[0];
                    revocationReason = reasonValue.valueBlock.valueDec;
                }
            }
            return CertStatus.createRevoked(revocationTime, revocationReason);
        }
        else if (asn1 instanceof asn1js.Primitive &&
            asn1.idBlock.tagClass === 3 &&
            asn1.idBlock.tagNumber === 2) {
            // unknown
            return CertStatus.createUnknown();
        }
        else {
            throw new Asn1ParseError('Invalid ASN.1 structure for CertStatus');
        }
    }
}
