// pki-lite/dist/ocsp/CertID.js
import { asn1js, PkiBase, derToAsn1 } from '../core/PkiBase.js';
import { AlgorithmIdentifier, } from '../algorithms/AlgorithmIdentifier.js';
import { OctetString } from '../asn1/OctetString.js';
import { Integer } from '../asn1/Integer.js';
import { Asn1ParseError } from '../core/errors/Asn1ParseError.js';
/**
 * ```asn
 * CertID ::= SEQUENCE {
 *     hashAlgorithm           AlgorithmIdentifier,
 *     issuerNameHash          OCTET STRING, -- Hash of issuer's DN
 *     issuerKeyHash           OCTET STRING, -- Hash of issuer's key
 *     serialNumber            CertificateSerialNumber
 * }
 * ```
 */
export class CertID extends PkiBase {
    hashAlgorithm;
    issuerNameHash;
    issuerKeyHash;
    serialNumber;
    constructor(options) {
        super();
        const { hashAlgorithm, issuerNameHash, issuerKeyHash, serialNumber } = options;
        this.hashAlgorithm = hashAlgorithm;
        this.issuerNameHash =
            issuerNameHash instanceof OctetString
                ? issuerNameHash
                : new OctetString({ bytes: issuerNameHash });
        this.issuerKeyHash =
            issuerKeyHash instanceof OctetString
                ? issuerKeyHash
                : new OctetString({ bytes: issuerKeyHash });
        this.serialNumber = new Integer({ value: serialNumber });
    }
    toAsn1() {
        return new asn1js.Sequence({
            value: [
                this.hashAlgorithm.toAsn1(),
                this.issuerNameHash.toAsn1(),
                this.issuerKeyHash.toAsn1(),
                this.serialNumber.toAsn1(),
            ],
        });
    }
    static fromAsn1(asn1) {
        if (!(asn1 instanceof asn1js.Sequence)) {
            throw new Asn1ParseError('Invalid ASN.1 structure: expected SEQUENCE');
        }
        const sequence = asn1;
        if (sequence.valueBlock.value.length !== 4) {
            throw new Asn1ParseError('Invalid CertID: expected 4 elements');
        }
        const hashAlgorithm = AlgorithmIdentifier.fromAsn1(sequence.valueBlock.value[0]);
        const issuerNameHash = OctetString.fromAsn1(sequence.valueBlock.value[1]);
        const issuerKeyHash = OctetString.fromAsn1(sequence.valueBlock.value[2]);
        const serialNumberBlock = sequence.valueBlock.value[3];
        const serialNumber = serialNumberBlock.valueBlock.valueHexView;
        return new CertID({
            hashAlgorithm,
            issuerNameHash,
            issuerKeyHash,
            serialNumber,
        });
    }
    static fromDer(der) {
        return CertID.fromAsn1(derToAsn1(der));
    }
    /**
     * Creates a CertID from issuer and subject certificates.
     *
     * @param issuerCertificate The issuer certificate
     * @param subjectCertificate The subject certificate
     * @param hashAlgorithm Optional hash algorithm (defaults to SHA-256)
     * @returns Promise resolving to the CertID
     */
    static async forCertificate(issuerCertificate, subjectCertificate, hashAlgorithm) {
        const hash = hashAlgorithm ?? AlgorithmIdentifier.digestAlgorithm('SHA-256');
        return new CertID({
            hashAlgorithm: hash,
            issuerNameHash: await hash.digest(issuerCertificate.tbsCertificate.subject.toDer()),
            issuerKeyHash: await hash.digest(issuerCertificate.tbsCertificate.subjectPublicKeyInfo.subjectPublicKey.toDer()),
            serialNumber: subjectCertificate.tbsCertificate.serialNumber,
        });
    }
}
