// pki-lite/dist/ocsp/OCSPSignature.js
import { asn1js, PkiBase, derToAsn1 } from '../core/PkiBase.js';
import { AlgorithmIdentifier } from '../algorithms/AlgorithmIdentifier.js';
import { Certificate } from '../x509/Certificate.js';
import { Asn1ParseError } from '../core/errors/Asn1ParseError.js';
/**
 * Signature ::= SEQUENCE {
 *     signatureAlgorithm   AlgorithmIdentifier,
 *     signature            BIT STRING,
 *     certs             [0] EXPLICIT SEQUENCE OF Certificate OPTIONAL
 * }
 */
export class OCSPSignature extends PkiBase {
    signatureAlgorithm;
    signature;
    certs;
    constructor(options) {
        super();
        this.signatureAlgorithm = options.signatureAlgorithm;
        this.signature = new Uint8Array(options.signature);
        this.certs = options.certs;
    }
    toAsn1() {
        const values = [
            this.signatureAlgorithm.toAsn1(),
            new asn1js.BitString({ valueHex: this.signature }),
        ];
        if (this.certs && this.certs.length > 0) {
            const certsSequence = new asn1js.Sequence({
                value: this.certs.map((cert) => cert.toAsn1()),
            });
            values.push(new asn1js.Constructed({
                idBlock: {
                    tagClass: 3, // CONTEXT-SPECIFIC
                    tagNumber: 0, // [0]
                },
                value: [certsSequence],
            }));
        }
        return new asn1js.Sequence({
            value: values,
        });
    }
    static fromAsn1(asn1) {
        if (!(asn1 instanceof asn1js.Sequence)) {
            throw new Asn1ParseError('Invalid ASN.1 structure: expected SEQUENCE');
        }
        const sequence = asn1;
        if (sequence.valueBlock.value.length < 2 ||
            sequence.valueBlock.value.length > 3) {
            throw new Asn1ParseError('Invalid Signature: expected 2 or 3 elements');
        }
        const signatureAlgorithm = AlgorithmIdentifier.fromAsn1(sequence.valueBlock.value[0]);
        const signatureBlock = sequence.valueBlock.value[1];
        const signature = signatureBlock.valueBlock.valueHexView;
        let certs = undefined;
        if (sequence.valueBlock.value.length === 3) {
            const certsContainer = sequence.valueBlock
                .value[2];
            if (certsContainer.idBlock.tagClass !== 3 ||
                certsContainer.idBlock.tagNumber !== 0) {
                throw new Asn1ParseError('Invalid Signature: expected [0] tag for certs');
            }
            const certsSequence = certsContainer.valueBlock
                .value[0];
            certs = certsSequence.valueBlock.value.map((cert) => Certificate.fromAsn1(cert));
        }
        return new OCSPSignature({
            signatureAlgorithm,
            signature,
            certs,
        });
    }
    static fromDer(der) {
        return OCSPSignature.fromAsn1(derToAsn1(der));
    }
}
