// pki-lite/dist/ocsp/SingleResponse.js
import { GeneralizedTime } from '../asn1/GeneralizedTime.js';
import { asn1js, PkiBase } from '../core/PkiBase.js';
import { Extension } from '../x509/Extension.js';
import { CertID } from './CertID.js';
import { CertStatus } from './CertStatus.js';
import { Asn1ParseError } from '../core/errors/Asn1ParseError.js';
/**
 * SingleResponse ::= SEQUENCE {
 *     certID                       CertID,
 *     certStatus                   CertStatus,
 *     thisUpdate                   GeneralizedTime,
 *     nextUpdate           [0]     EXPLICIT GeneralizedTime OPTIONAL,
 *     singleExtensions     [1]     EXPLICIT Extensions OPTIONAL
 * }
 */
export class SingleResponse extends PkiBase {
    certID;
    certStatus;
    thisUpdate;
    nextUpdate;
    singleExtensions;
    constructor(options) {
        super();
        this.certID = options.certID;
        this.certStatus = options.certStatus;
        this.thisUpdate = new GeneralizedTime({ time: options.thisUpdate });
        this.nextUpdate = options.nextUpdate
            ? new GeneralizedTime({ time: options.nextUpdate })
            : undefined;
        this.singleExtensions = options.singleExtensions;
    }
    toAsn1() {
        const values = [
            this.certID.toAsn1(),
            this.certStatus.toAsn1(),
            this.thisUpdate.toAsn1(),
        ];
        if (this.nextUpdate) {
            values.push(new asn1js.Constructed({
                idBlock: {
                    tagClass: 3, // CONTEXT-SPECIFIC
                    tagNumber: 0, // [0]
                },
                value: [this.nextUpdate.toAsn1()],
            }));
        }
        if (this.singleExtensions && this.singleExtensions.length > 0) {
            values.push(new asn1js.Constructed({
                idBlock: {
                    tagClass: 3, // CONTEXT-SPECIFIC
                    tagNumber: 1, // [1]
                },
                value: [
                    new asn1js.Sequence({
                        value: this.singleExtensions.map((ext) => ext.toAsn1()),
                    }),
                ],
            }));
        }
        return new asn1js.Sequence({
            value: values,
        });
    }
    static fromAsn1(asn1) {
        if (!(asn1 instanceof asn1js.Sequence)) {
            throw new Asn1ParseError('Invalid ASN.1 structure: expected SEQUENCE');
        }
        const sequence = asn1;
        if (sequence.valueBlock.value.length < 3) {
            throw new Asn1ParseError('Invalid SingleResponse: expected at least 3 elements');
        }
        const certID = CertID.fromAsn1(sequence.valueBlock.value[0]);
        const certStatus = CertStatus.fromAsn1(sequence.valueBlock.value[1]);
        const thisUpdateBlock = sequence.valueBlock
            .value[2];
        const thisUpdate = thisUpdateBlock.toDate();
        let nextUpdate = undefined;
        let singleExtensions = undefined;
        let currentIndex = 3;
        while (currentIndex < sequence.valueBlock.value.length) {
            const block = sequence.valueBlock.value[currentIndex];
            if (block.idBlock.tagClass === 3 &&
                block.idBlock.tagNumber === 0 &&
                !nextUpdate) {
                // nextUpdate
                const nextUpdateBlock = block.valueBlock
                    .value[0];
                nextUpdate = nextUpdateBlock.toDate();
            }
            else if (block.idBlock.tagClass === 3 &&
                block.idBlock.tagNumber === 1 &&
                !singleExtensions) {
                // singleExtensions
                const extensionsSequence = block.valueBlock
                    .value[0];
                singleExtensions = extensionsSequence.valueBlock.value.map((ext) => Extension.fromAsn1(ext));
            }
            currentIndex++;
        }
        return new SingleResponse({
            certID,
            certStatus,
            thisUpdate,
            nextUpdate,
            singleExtensions,
        });
    }
}
