// pdf-lite/dist/acroform/pdf-acro-form.js
import { PdfDictionary } from '../core/objects/pdf-dictionary.js';
import { PdfArray } from '../core/objects/pdf-array.js';
import { PdfString } from '../core/objects/pdf-string.js';
import { PdfObjectReference } from '../core/objects/pdf-object-reference.js';
import { PdfIndirectObject } from '../core/objects/pdf-indirect-object.js';
import { PdfBoolean } from '../core/objects/pdf-boolean.js';
import { PdfNumber } from '../core/objects/pdf-number.js';
import { PdfFormField } from './fields/pdf-form-field.js';
// Import subclasses to trigger static registration blocks
import './fields/pdf-text-form-field.js';
import './fields/pdf-button-form-field.js';
import './fields/pdf-choice-form-field.js';
import './fields/pdf-signature-form-field.js';
import { buildEncodingMap } from '../utils/decodeWithFontEncoding.js';
import { PdfXfaForm } from './xfa/pdf-xfa-form.js';
export class PdfAcroForm extends PdfIndirectObject {
    constructor(options) {
        super(options ??
            new PdfIndirectObject({
                content: new PdfDictionary(),
            }));
    }
    get needAppearances() {
        return (this.content.get('NeedAppearances')?.as(PdfBoolean)?.value ?? false);
    }
    set needAppearances(value) {
        this.content.set('NeedAppearances', new PdfBoolean(value));
    }
    get signatureFlags() {
        return this.content.get('SigFlags')?.as(PdfNumber)?.value ?? 0;
    }
    set signatureFlags(flags) {
        this.content.set('SigFlags', new PdfNumber(flags));
    }
    get defaultAppearance() {
        return this.content.get('DA')?.as(PdfString)?.value ?? null;
    }
    set defaultAppearance(da) {
        this.content.set('DA', new PdfString(da));
    }
    get defaultQuadding() {
        return this.content.get('Q')?.as(PdfNumber)?.value ?? 0;
    }
    set defaultQuadding(q) {
        this.content.set('Q', new PdfNumber(q));
    }
    get defaultResources() {
        return this.content.get('DR')?.as(PdfDictionary) ?? null;
    }
    set defaultResources(resources) {
        if (resources === null) {
            this.content.delete('DR');
        }
        else {
            this.content.set('DR', resources);
        }
    }
    get fields() {
        const result = [];
        const seen = new Set();
        const collect = (refs) => {
            for (const ref of refs) {
                const key = ref.key;
                if (seen.has(key))
                    continue;
                seen.add(key);
                const resolved = ref.resolve();
                if (!(resolved?.content instanceof PdfDictionary))
                    continue;
                const dict = resolved.content;
                const tEntry = dict.get('T');
                const tValue = tEntry instanceof PdfString ? tEntry.value : null;
                const hasFt = dict.has('FT');
                if (tValue) {
                    const field = PdfFormField.create(resolved);
                    field.form = this;
                    result.push(field);
                }
                // Only recurse into Kids for non-terminal (group) fields — those
                // without FT. Terminal fields may have Kids that are pure widget
                // annotations (no T) which would produce duplicate names.
                if (!hasFt) {
                    const kids = dict.get('Kids');
                    if (kids instanceof PdfArray && kids.items.length > 0) {
                        collect(kids.items);
                    }
                }
            }
        };
        const fieldsRaw = this.content.get('Fields');
        const resolvedFields = fieldsRaw instanceof PdfObjectReference
            ? fieldsRaw.resolve()?.content
            : fieldsRaw;
        collect(resolvedFields instanceof PdfArray
            ? resolvedFields.items
            : []);
        return result;
    }
    addField(...fields) {
        const content = this.content;
        const fieldsRaw = content.get('Fields');
        let fieldsArray = fieldsRaw instanceof PdfObjectReference
            ? fieldsRaw.resolve()?.content
            : fieldsRaw;
        if (!fieldsArray) {
            fieldsArray = new PdfArray();
            content.set('Fields', fieldsArray);
        }
        for (const field of fields) {
            fieldsArray.items.push(field.reference);
            this._addWidgetToPage(field);
            // Also register any child widgets (e.g. radio button group kids)
            for (const child of field.children) {
                this._addWidgetToPage(child);
            }
        }
    }
    _addWidgetToPage(field) {
        const page = field.page;
        if (!page)
            return;
        const annots = page.annotations;
        // Unregistered fields all share key "-1/0", so compare by object
        // identity (via resolve()) rather than by key.
        const alreadyPresent = annots.items.some((r) => {
            if (!(r instanceof PdfObjectReference))
                return false;
            try {
                return r.resolve() === field;
            }
            catch {
                return false;
            }
        });
        if (!alreadyPresent) {
            annots.items.push(field.reference);
        }
    }
    set fields(newFields) {
        this.content.set('Fields', PdfArray.refs(newFields));
    }
    setValues(values) {
        for (const field of this.fields) {
            const name = field.name;
            if (name in values && values[name] !== undefined) {
                const result = this._fireValidate(field, values[name]);
                if (!result.rc)
                    continue;
                field.value = result.value;
            }
        }
        this._fireCalculate();
    }
    importData(fields) {
        for (const field of this.fields) {
            const name = field.name;
            if (name && name in fields) {
                const result = this._fireValidate(field, fields[name]);
                if (!result.rc)
                    continue;
                field.value = result.value;
            }
        }
        this._fireCalculate();
    }
    _fireValidate(field, value) {
        const validateAction = field.actions?.validate;
        if (!validateAction?.code)
            return { rc: true, value };
        const event = {
            fieldName: field.name,
            value,
            willCommit: true,
            rc: true,
        };
        validateAction.execute(event);
        return { rc: event.rc, value: event.value };
    }
    _fireCalculate() {
        const co = this.content.get('CO');
        if (!(co instanceof PdfArray))
            return;
        const allFields = this.fields;
        const fieldsByObjNum = new Map();
        for (const field of allFields) {
            if (field.objectNumber != null) {
                fieldsByObjNum.set(field.objectNumber, field);
            }
        }
        for (const item of co.items) {
            if (!(item instanceof PdfObjectReference))
                continue;
            const resolved = item.resolve();
            if (!resolved)
                continue;
            const field = fieldsByObjNum.get(resolved.objectNumber);
            if (!field)
                continue;
            const calcAction = field.actions?.calculate;
            if (!calcAction?.code)
                continue;
            const event = {
                fieldName: field.name,
                value: field.value,
                rc: true,
            };
            calcAction.execute(event);
            if (event.rc && event.value !== field.value) {
                field.value = event.value;
            }
        }
    }
    exportData() {
        const result = {};
        for (const field of this.fields) {
            const name = field.name;
            if (name) {
                result[name] = field.value;
            }
        }
        return result;
    }
    jsEngine;
    fontEncodingMaps = new Map();
    getFontEncodingMap(fontName) {
        if (this.fontEncodingMaps.has(fontName)) {
            return this.fontEncodingMaps.get(fontName);
        }
        const dr = this.defaultResources;
        const fontRaw = dr?.get('Font');
        let fontDict;
        if (fontRaw instanceof PdfObjectReference) {
            const resolved = fontRaw.resolve()?.content;
            if (resolved instanceof PdfDictionary)
                fontDict = resolved;
        }
        else if (fontRaw instanceof PdfDictionary) {
            fontDict = fontRaw;
        }
        if (!fontDict)
            return null;
        const fontEntry = fontDict.get(fontName);
        if (!fontEntry)
            return null;
        let fontObj;
        if (fontEntry instanceof PdfObjectReference) {
            const resolved = fontEntry.resolve();
            if (resolved?.content instanceof PdfDictionary) {
                fontObj = resolved.content;
            }
        }
        else if (fontEntry instanceof PdfDictionary) {
            fontObj = fontEntry;
        }
        if (!fontObj)
            return null;
        let encObj = fontObj.get('Encoding');
        if (encObj instanceof PdfObjectReference) {
            encObj = encObj.resolve()?.content;
        }
        const encDict = encObj instanceof PdfDictionary ? encObj : undefined;
        const diffs = encDict?.get('Differences')?.as(PdfArray);
        if (!diffs)
            return null;
        const map = buildEncodingMap(diffs);
        if (map) {
            this.fontEncodingMaps.set(fontName, map);
        }
        return map;
    }
    get xfa() {
        const xfaEntry = this.content.get('XFA');
        if (!xfaEntry)
            return null;
        if (!(xfaEntry instanceof PdfArray)) {
            return null;
        }
        return new PdfXfaForm(xfaEntry);
    }
}
