// pdf-lite/dist/security/crypt-filters/base.js
import { PdfDictionary } from '../../core/objects/pdf-dictionary.js';
import { PdfName } from '../../core/objects/pdf-name.js';
import { PdfNumber } from '../../core/objects/pdf-number.js';
/**
 * Abstract base class for PDF crypt filters.
 * Crypt filters define how specific types of data (strings, streams, files) are encrypted.
 *
 * @example
 * ```typescript
 * const filter = new AesV2CryptFilter({ authEvent: 'DocOpen' })
 * filter.setSecurityHandler(securityHandler)
 * const encrypted = await filter.encrypt(data, objectNumber, generationNumber)
 * ```
 */
export class PdfCryptFilter {
    /** Crypt filter method identifier. */
    cfm;
    /** Authentication event trigger. */
    authEvent;
    /** Key length in bits. */
    length;
    /** Security handler for cryptographic operations. */
    securityHandler;
    /**
     * Creates a new crypt filter with the specified options.
     *
     * @param options - Configuration options for the crypt filter.
     */
    constructor(options) {
        const { cfm, authEvent, length, securityHandler } = options;
        this.securityHandler = securityHandler;
        this.cfm = cfm;
        this.authEvent = authEvent;
        this.length = length ?? 0;
    }
    /**
     * Sets the security handler for this crypt filter.
     *
     * @param handler - The security handler to use for key derivation.
     */
    setSecurityHandler(handler) {
        this.securityHandler = handler;
    }
    /**
     * Gets the current security handler.
     *
     * @returns The security handler, or undefined if not set.
     */
    getSecurityHandler() {
        return this.securityHandler;
    }
    /**
     * Encrypts data using this crypt filter.
     *
     * @param data - The data to encrypt.
     * @param objectNumber - The PDF object number for key derivation.
     * @param generationNumber - The PDF generation number for key derivation.
     * @returns The encrypted data.
     */
    async encrypt(data, objectNumber, generationNumber) {
        const cipher = await this.getCipher(objectNumber, generationNumber);
        return cipher.encrypt(data);
    }
    /**
     * Decrypts data using this crypt filter.
     *
     * @param data - The data to decrypt.
     * @param objectNumber - The PDF object number for key derivation.
     * @param generationNumber - The PDF generation number for key derivation.
     * @returns The decrypted data.
     */
    async decrypt(data, objectNumber, generationNumber) {
        const cipher = await this.getCipher(objectNumber, generationNumber);
        return cipher.decrypt(data);
    }
    /**
     * Converts this crypt filter to a PDF dictionary representation.
     *
     * @returns The crypt filter dictionary.
     */
    toDictionary() {
        const dict = new PdfDictionary({
            ...(this.authEvent
                ? { AuthEvent: new PdfName(this.authEvent) }
                : {}),
            CFM: new PdfName(this.cfm),
            ...(this.length
                ? {
                    Length: new PdfNumber(this.length),
                }
                : {}),
        });
        return dict;
    }
}
