// pdf-lite/dist/acroform/fields/pdf-form-field-flags.js
import { PdfNumber } from '../../core/objects/pdf-number.js';
/**
 * Field-specific Ff flag accessors for form fields.
 * These are separate from annotation flags (F field).
 * Extends PdfNumber so it can be stored directly in a PDF dictionary.
 */
export class PdfFormFieldFlags extends PdfNumber {
    constructor(value = 0) {
        super(value);
        this.preTokens = typeof value === 'number' ? undefined : value.preTokens;
        this.postTokens =
            typeof value === 'number' ? undefined : value.postTokens;
    }
    get flags() {
        return this.value;
    }
    set flags(v) {
        this.value = v instanceof PdfFormFieldFlags ? v.value : v;
    }
    getFlag(bit) {
        return (this.flags & bit) !== 0;
    }
    setFlag(bit, val) {
        this.flags = val ? this.flags | bit : this.flags & ~bit;
    }
    get readOnly() {
        return this.getFlag(1);
    }
    set readOnly(v) {
        this.setFlag(1, v);
    }
    get required() {
        return this.getFlag(2);
    }
    set required(v) {
        this.setFlag(2, v);
    }
    get noExport() {
        return this.getFlag(4);
    }
    set noExport(v) {
        this.setFlag(4, v);
    }
    get multiline() {
        return this.getFlag(4096);
    }
    set multiline(v) {
        this.setFlag(4096, v);
    }
    get password() {
        return this.getFlag(8192);
    }
    set password(v) {
        this.setFlag(8192, v);
    }
    get noToggleToOff() {
        return this.getFlag(16384);
    }
    set noToggleToOff(v) {
        this.setFlag(16384, v);
    }
    get radio() {
        return this.getFlag(32768);
    }
    set radio(v) {
        this.setFlag(32768, v);
    }
    get pushButton() {
        return this.getFlag(65536);
    }
    set pushButton(v) {
        this.setFlag(65536, v);
    }
    get combo() {
        return this.getFlag(131072);
    }
    set combo(v) {
        this.setFlag(131072, v);
    }
    get edit() {
        return this.getFlag(262144);
    }
    set edit(v) {
        this.setFlag(262144, v);
    }
    get sort() {
        return this.getFlag(524288);
    }
    set sort(v) {
        this.setFlag(524288, v);
    }
    get multiSelect() {
        return this.getFlag(2097152);
    }
    set multiSelect(v) {
        this.setFlag(2097152, v);
    }
    get doNotSpellCheck() {
        return this.getFlag(4194304);
    }
    set doNotSpellCheck(v) {
        this.setFlag(4194304, v);
    }
    get doNotScroll() {
        return this.getFlag(8388608);
    }
    set doNotScroll(v) {
        this.setFlag(8388608, v);
    }
    get comb() {
        return this.getFlag(16777216);
    }
    set comb(v) {
        this.setFlag(16777216, v);
    }
    get commitOnSelChange() {
        return this.getFlag(67108864);
    }
    set commitOnSelChange(v) {
        this.setFlag(67108864, v);
    }
}
