// pki-lite/dist/asn1/OctetString.js
import { asn1js, derToAsn1, PkiBase, PkiSequence, PkiSet, } from '../core/PkiBase.js';
import { Asn1ParseError } from '../core/errors/Asn1ParseError.js';
/**
 * Represents an ASN.1 OCTET STRING value.
 *
 * @asn
 * ```asn
 * OctetString ::= <value>
 * ```
 */
export class OctetString extends PkiBase {
    bytes;
    constructor(options) {
        super();
        const { bytes } = options;
        if (bytes instanceof OctetString) {
            this.bytes = bytes.bytes;
        }
        else if (typeof bytes === 'string') {
            this.bytes = new TextEncoder().encode(bytes);
        }
        else if (bytes instanceof PkiBase ||
            bytes instanceof PkiSequence ||
            bytes instanceof PkiSet) {
            this.bytes = bytes.toDer();
        }
        else {
            this.bytes = new Uint8Array(bytes);
        }
    }
    toAsn1() {
        return new asn1js.OctetString({ valueHex: this.bytes });
    }
    toUint8Array() {
        return this.bytes;
    }
    static fromAsn1(asn1) {
        if (!(asn1 instanceof asn1js.OctetString) &&
            !(asn1 instanceof asn1js.Primitive)) {
            throw new Asn1ParseError('Invalid ASN.1 structure: expected OCTET STRING but got ' +
                asn1.constructor.name);
        }
        return new OctetString({ bytes: asn1.valueBlock.valueHexView });
    }
    static fromDer(der) {
        return OctetString.fromAsn1(derToAsn1(der));
    }
    parseAs(type) {
        if (type.fromDer) {
            return type.fromDer(this.bytes);
        }
        else if (type.fromAsn1) {
            return type.fromAsn1(derToAsn1(this.bytes));
        }
        else {
            throw new Asn1ParseError('Invalid type: must implement fromAsn1 or fromDer');
        }
    }
}
