// pdf-lite/dist/core/objects/pdf-number.js
import { Ref } from '../ref.js';
import { PdfNumberToken } from '../tokens/number-token.js';
import { PdfObject } from './pdf-object.js';
export class PdfNumber extends PdfObject {
    #value;
    padTo;
    decimalPlaces;
    isByteOffset = false;
    constructor(options = 0) {
        super();
        if (options instanceof Ref) {
            this.#value = options;
            this.padTo = 0;
            this.decimalPlaces = 0;
        }
        else if (options instanceof PdfNumber) {
            this.#value = options.ref;
            this.padTo = options.padTo;
            this.decimalPlaces = options.decimalPlaces;
            this.preTokens = options.preTokens;
            this.postTokens = options.postTokens;
        }
        else if (typeof options === 'number') {
            this.#value = new Ref(options);
            this.padTo = 0;
            this.decimalPlaces = 0;
        }
        else {
            this.#value =
                typeof options.value === 'number'
                    ? new Ref(options.value)
                    : options.value;
            this.padTo = options.padTo ?? 0;
            this.decimalPlaces = options.decimalPlaces ?? 0;
        }
        if (this.value === Number.POSITIVE_INFINITY ||
            this.value === Number.NEGATIVE_INFINITY) {
            throw new Error('PDF numbers cannot be infinite');
        }
    }
    get ref() {
        return this.#value;
    }
    get value() {
        return this.#value.resolve();
    }
    set value(value) {
        this.#value.update(value);
    }
    onChange(callback) {
        this.#value.onUpdate(callback);
    }
    toToken() {
        return new PdfNumberToken({
            value: this.#value,
            padTo: this.padTo,
            decimalPlaces: this.decimalPlaces,
            isByteToken: this.isByteOffset,
        });
    }
    tokenize() {
        return [this.toToken()];
    }
    cloneImpl() {
        const cloned = new PdfNumber({
            value: this.value,
            padTo: this.padTo,
            decimalPlaces: this.decimalPlaces,
        });
        return cloned;
    }
    toJSON() {
        return { type: 'number', value: this.value };
    }
    setModified(modified) {
        super.setModified(modified);
        this.#value.setModified(modified);
    }
    isModified() {
        return super.isModified() || this.#value.isModified();
    }
}
export class PdfByteOffsetNumber extends PdfNumber {
    isByteOffset = true;
}
