// pdf-lite/dist/core/objects/pdf-array.js
import { needsCentralWhitespace } from '../../utils/needsCentralWhitespace.js';
import { PdfEndArrayToken } from '../tokens/end-array-token.js';
import { PdfStartArrayToken } from '../tokens/start-array-token.js';
import { PdfWhitespaceToken } from '../tokens/whitespace-token.js';
import { PdfIndirectObject } from './pdf-indirect-object.js';
import { PdfObject } from './pdf-object.js';
export class PdfArray extends PdfObject {
    items;
    innerTokens = [];
    constructor(items = []) {
        super();
        this.items = items;
    }
    static refs(items) {
        return new PdfArray(items.map((item) => item.reference));
    }
    get length() {
        return this.items.length;
    }
    push(item) {
        this.items.push(item);
    }
    toJSON() {
        return { type: 'array', items: this.items.map((item) => item.toJSON()) };
    }
    get isTrailingDelimited() {
        return true;
    }
    tokenize() {
        const items = this.items.flatMap((item, index) => {
            const tokens = item.toTokens();
            const preTokens = item.preTokens
                ? []
                : index === 0
                    ? [PdfWhitespaceToken.SPACE]
                    : [];
            const postTokens = item.postTokens ? [] : [PdfWhitespaceToken.SPACE];
            if (index !== this.items.length - 1 &&
                postTokens.length === 0 &&
                needsCentralWhitespace(item, this.items[index + 1])) {
                postTokens.push(PdfWhitespaceToken.SPACE);
            }
            return [...preTokens, ...tokens, ...postTokens];
        });
        return [
            new PdfStartArrayToken(),
            ...this.innerTokens,
            ...items,
            new PdfEndArrayToken(),
        ];
    }
    cloneImpl() {
        const cloned = new PdfArray(this.items.map((x) => x.clone()));
        return cloned;
    }
    setModified(modified) {
        super.setModified(modified);
        this.items.forEach((item) => item.setModified(modified));
    }
    isModified() {
        return (super.isModified() || this.items.some((item) => item.isModified()));
    }
    setImmutable(immutable) {
        super.setImmutable(immutable);
        this.items.forEach((item) => item.setImmutable(immutable));
    }
    refs() {
        const refs = this.items.map((item) => {
            if (item instanceof PdfIndirectObject) {
                return item.reference;
            }
            else {
                throw new Error('Cannot get reference of non-indirect object in array');
            }
        });
        return new PdfArray(refs);
    }
    [Symbol.iterator]() {
        return this.items[Symbol.iterator]();
    }
}
