// pdf-lite/dist/core/objects/pdf-object.js
import { concatUint8Arrays } from '../../utils/concatUint8Arrays.js';
import { PdfWhitespaceToken } from '../tokens/whitespace-token.js';
export class PdfObject {
    /** Optional tokens to prepend or append during serialization */
    preTokens;
    /** Optional tokens to prepend or append during serialization */
    postTokens;
    /** Indicates whether the object has been modified. By default, assume it has been modified because it's a new object */
    modified = true;
    /** Indicates whether the object is immutable (cannot be modified) */
    immutable = false;
    /** Cached byte representation of the object, if available */
    cachedTokens;
    /** The type of this PDF object */
    get objectType() {
        return this.constructor.name;
    }
    /**
     * Returns true if this object's serialized form ends with a self-delimiting
     * character (e.g., `)`, `>`, `]`, `>>`). Such objects do not require trailing
     * whitespace before the next token.
     */
    get isTrailingDelimited() {
        return false;
    }
    /** Indicates whether the object has been modified. Override this method if the modified state is determined differently */
    isModified() {
        return this.modified;
    }
    /** Sets the modified state of the object. Override this method if the modified state is determined differently */
    setModified(modified = true) {
        this.modified = modified;
    }
    /** Indicates whether the object is immutable (cannot be modified) */
    isImmutable() {
        return this.immutable;
    }
    /** Sets the immutable state of the object */
    setImmutable(immutable = true) {
        if (immutable === this.immutable) {
            return;
        }
        this.immutable = immutable;
        if (immutable) {
            this.cachedTokens = this.toTokens();
        }
        else {
            this.cachedTokens = undefined;
        }
    }
    /** Converts the object to an array of PdfTokens, including any pre or post tokens */
    toTokens() {
        if (this.cachedTokens) {
            return this.cachedTokens;
        }
        return [
            ...(this.preTokens ?? []),
            ...this.tokenize(),
            ...(this.postTokens ?? []),
        ];
    }
    /** Converts the object to a ByteArray, optionally padding to a specified length */
    toBytes(padTo) {
        const tokens = this.toTokens();
        const byteArray = concatUint8Arrays(tokens.map((token) => token.toBytes()));
        if (padTo) {
            const paddedArray = new Uint8Array(padTo);
            paddedArray.set(byteArray);
            return paddedArray;
        }
        return byteArray;
    }
    /** Converts the object to a string representation */
    toString() {
        let str = '';
        for (const byte of this.toBytes()) {
            str += String.fromCharCode(byte);
        }
        return str;
    }
    /** Attempts to cast the object to a specific PdfObject subclass */
    as(ctor) {
        if (this instanceof ctor) {
            return this;
        }
        throw new Error(`Cannot cast object of type ${this.objectType} to ${ctor.name}`);
    }
    /** Creates a deep clone of the object */
    clone() {
        const cloned = this.cloneImpl();
        cloned.setModified(this.modified);
        return cloned;
    }
    /** Compares this object to another for equality based on their token representations */
    equals(other) {
        if (!other) {
            return false;
        }
        const tokensA = this.toTokens().filter((x) => !(x instanceof PdfWhitespaceToken));
        const tokensB = other
            .toTokens()
            .filter((x) => !(x instanceof PdfWhitespaceToken));
        if (tokensA.length !== tokensB.length) {
            return false;
        }
        for (let i = 0; i < tokensA.length; i++) {
            const tokenA = tokensA[i];
            const tokenB = tokensB[i];
            if (!tokenA.equals(tokenB)) {
                return false;
            }
        }
        return true;
    }
    /**
     * Serializes the document to a Base64-encoded string.
     *
     * @returns A promise that resolves to the PDF document as a Base64 string
     */
    toBase64() {
        const bytes = this.toBytes();
        let binary = '';
        const len = bytes.byteLength;
        for (let i = 0; i < len; i++) {
            binary += String.fromCharCode(bytes[i]);
        }
        return btoa(binary);
    }
}
