// pdf-lite/dist/core/objects/pdf-trailer.js
import { Ref } from '../ref.js';
import { PdfByteOffsetToken } from '../tokens/byte-offset-token.js';
import { PdfTrailerToken } from '../tokens/trailer-token.js';
import { PdfWhitespaceToken } from '../tokens/whitespace-token.js';
import { PdfDictionary } from './pdf-dictionary.js';
import { PdfObject } from './pdf-object.js';
export class PdfTrailer extends PdfObject {
    dict;
    offset = new Ref(0);
    constructor(entries) {
        super();
        this.dict =
            entries instanceof PdfDictionary
                ? entries
                : new PdfDictionary(entries);
    }
    toJSON() {
        return { type: 'trailer', dict: this.dict.toJSON() };
    }
    tokenize() {
        return [
            new PdfByteOffsetToken(this.offset),
            new PdfTrailerToken(),
            ...(this.dict.preTokens ? [] : [PdfWhitespaceToken.NEWLINE]),
            ...this.dict.toTokens(),
        ];
    }
    cloneImpl() {
        const cloned = new PdfTrailer(this.dict.clone());
        return cloned;
    }
    setModified(modified) {
        super.setModified(modified);
        this.dict.setModified(modified);
        this.offset.setModified(modified);
    }
    isModified() {
        return (super.isModified() || this.dict.isModified() || this.offset.modified);
    }
}
