// pdf-lite/dist/core/objects/pdf-hexadecimal.js
import { bytesToHexBytes } from '../../utils/bytesToHexBytes.js';
import { bytesToString } from '../../utils/bytesToString.js';
import { hexBytesToBytes } from '../../utils/hexBytesToBytes.js';
import { stringToBytes } from '../../utils/stringToBytes.js';
import { PdfHexadecimalToken } from '../tokens/hexadecimal-token.js';
import { PdfObject } from './pdf-object.js';
export class PdfHexadecimal extends PdfObject {
    /**
     * The raw byte value represented by this hexadecimal object.
     * NB: This is  the hexadecimal representation, not the actual byte values.
     */
    _raw;
    /**
     * Original bytes from the PDF file, including angle brackets.
     * Used to preserve exact formatting for incremental updates.
     */
    _originalBytes;
    constructor(value, format = 'hex', originalBytes) {
        super();
        let bytes;
        if (format === 'bytes') {
            bytes = bytesToHexBytes(value instanceof Uint8Array ? value : stringToBytes(value));
        }
        else {
            bytes = value instanceof Uint8Array ? value : stringToBytes(value);
        }
        this._raw = bytes;
        this._originalBytes = originalBytes;
    }
    get raw() {
        return this._raw;
    }
    set raw(raw) {
        this.setModified();
        this._raw = raw;
        // Clear original bytes when modified
        this._originalBytes = undefined;
    }
    static toHexadecimal(data) {
        return new PdfHexadecimal(data, 'bytes');
    }
    get bytes() {
        return hexBytesToBytes(this.raw);
    }
    toHexBytes() {
        return bytesToHexBytes(this.raw);
    }
    toHexString() {
        return bytesToString(this.toHexBytes());
    }
    toJSON() {
        return { type: 'hexadecimal', value: this.toHexString() };
    }
    get isTrailingDelimited() {
        return true;
    }
    tokenize() {
        return [new PdfHexadecimalToken(this.raw, this._originalBytes)];
    }
    cloneImpl() {
        const cloned = new PdfHexadecimal(new Uint8Array(this.raw), 'hex', this._originalBytes
            ? new Uint8Array(this._originalBytes)
            : undefined);
        return cloned;
    }
}
