// pdf-lite/dist/core/objects/pdf-indirect-object.js
import { PdfEndObjectToken } from '../tokens/end-object-token.js';
import { PdfStartObjectToken } from '../tokens/start-object-token.js';
import { PdfWhitespaceToken } from '../tokens/whitespace-token.js';
import { PdfNull } from './pdf-null.js';
import { PdfObject } from './pdf-object.js';
import { PdfObjectReference } from './pdf-object-reference.js';
import { PdfStream } from './pdf-stream.js';
import { Ref } from '../ref.js';
import { PdfByteOffsetToken } from '../tokens/byte-offset-token.js';
export class PdfIndirectObject extends PdfObject {
    static MAX_ORDER_INDEX = 2147483647;
    objectNumber;
    generationNumber;
    content;
    offset;
    encryptable;
    compressed;
    orderIndex;
    constructor(options) {
        if (options instanceof PdfIndirectObject) {
            super();
            this.objectNumber = options.objectNumber;
            this.generationNumber = options.generationNumber;
            this.content = options.content;
            this.offset = options.offset;
            this.compressed = options.compressed;
            return;
        }
        if (options instanceof PdfObject) {
            super();
            this.objectNumber = -1;
            this.generationNumber = 0;
            this.content = options;
            this.offset = new Ref(0);
            return;
        }
        super();
        this.objectNumber = options?.objectNumber ?? -1;
        this.generationNumber = options?.generationNumber ?? 0;
        this.content = options?.content ?? new PdfNull();
        this.offset =
            options?.offset instanceof Ref
                ? options.offset
                : new Ref(options?.offset ?? 0);
        this.encryptable = options?.encryptable;
        this.compressed = options?.compressed;
    }
    get reference() {
        const original = this;
        return new Proxy(new PdfObjectReference(this.objectNumber, this.generationNumber), {
            get: (_target, prop) => {
                // Always read from the live indirect object so objectNumber
                // stays current even after the object is added to a document.
                if (prop === 'resolve') {
                    return original.resolve.bind(original);
                }
                else if (prop === 'resolver') {
                    // Signal that this proxy reference is resolvable
                    // so collectMissingReferences can discover it
                    return {
                        resolve: () => original,
                    };
                }
                else if (prop === 'objectNumber') {
                    return original.objectNumber;
                }
                else if (prop === 'generationNumber') {
                    return original.generationNumber;
                }
                const value = new PdfObjectReference(original.objectNumber, original.generationNumber);
                return value[prop];
            },
        });
    }
    isEncryptable() {
        if (this.encryptable === undefined) {
            if (this.content instanceof PdfStream &&
                this.content.isType('XRef')) {
                return false;
            }
        }
        return this.encryptable ?? true;
    }
    static createPlaceholder(objectNumber, generationNumber, content) {
        return new PdfIndirectObject({
            objectNumber: objectNumber ?? -1,
            generationNumber: generationNumber ?? 0,
            content: content ?? new PdfNull(),
            offset: -1,
        });
    }
    inPdf() {
        return this.objectNumber >= 1;
    }
    matchesReference(ref) {
        if (!ref) {
            return false;
        }
        return (this.objectNumber === ref.objectNumber &&
            this.generationNumber === ref.generationNumber);
    }
    tokenize() {
        return [
            new PdfByteOffsetToken(this.offset),
            new PdfStartObjectToken(this.objectNumber, this.generationNumber),
            ...(this.content.preTokens ? [] : [PdfWhitespaceToken.NEWLINE]),
            ...this.content.toTokens(),
            ...(this.content.postTokens ? [] : [PdfWhitespaceToken.NEWLINE]),
            new PdfEndObjectToken(),
        ];
    }
    copyFrom(other) {
        this.objectNumber = other.objectNumber;
        this.generationNumber = other.generationNumber;
        this.content = other.content.clone();
        this.offset = other.offset.clone();
        this.modified = true;
    }
    cloneImpl() {
        return new PdfIndirectObject({
            objectNumber: this.objectNumber,
            generationNumber: this.generationNumber,
            content: this.content.clone(),
            offset: this.offset.resolve(),
            encryptable: this.encryptable,
            compressed: this.compressed,
        });
    }
    toJSON() {
        return {
            type: 'indirect-object',
            objectNumber: this.objectNumber,
            generationNumber: this.generationNumber,
            content: this.content.toJSON(),
        };
    }
    order() {
        return this.orderIndex ?? 0;
    }
    setModified(modified = true) {
        super.setModified(modified);
        this.content.setModified(modified);
        this.offset.setModified(modified);
    }
    isModified() {
        return (super.isModified() ||
            this.content.isModified() ||
            this.offset.isModified());
    }
    setImmutable(immutable) {
        super.setImmutable(immutable);
        this.content.setImmutable(immutable);
        this.offset.setImmutable(immutable);
    }
    becomes(cls) {
        if (this instanceof cls) {
            return this;
        }
        // Preserve identity-critical properties that subclass constructors
        // may inadvertently reset (e.g. PdfFont's super() creates a fresh
        // PdfIndirectObject with objectNumber = -1, and PdfFormField's
        // super() chain creates a fresh empty PdfDictionary as content).
        const savedObjectNumber = this.objectNumber;
        const savedGenerationNumber = this.generationNumber;
        const savedOffset = this.offset;
        const savedContent = this.content;
        const savedModified = this.modified;
        const savedEncryptable = this.encryptable;
        const savedCompressed = this.compressed;
        const newObject = new cls(this);
        Object.setPrototypeOf(this, cls.prototype);
        Object.assign(this, newObject);
        this.objectNumber = savedObjectNumber;
        this.generationNumber = savedGenerationNumber;
        this.offset = savedOffset;
        this.content = savedContent;
        this.modified = savedModified;
        this.encryptable = savedEncryptable;
        this.compressed = savedCompressed;
        return this;
    }
    resolve(cls) {
        if (cls) {
            return this.becomes(cls);
        }
        return this;
    }
    get key() {
        return `${this.objectNumber}/${this.generationNumber}`;
    }
}
