// pdf-lite/dist/core/objects/pdf-comment.js
import { stringToBytes } from '../../utils/stringToBytes.js';
import { PdfCommentToken } from '../tokens/comment-token.js';
import { PdfWhitespaceToken } from '../tokens/whitespace-token.js';
import { PdfObject } from './pdf-object.js';
export class PdfComment extends PdfObject {
    static EOF = new PdfComment('%EOF');
    /**
     * The comment value as raw bytes.
     */
    raw;
    constructor(raw) {
        super();
        this.raw = typeof raw === 'string' ? stringToBytes(raw) : raw;
    }
    isEof() {
        return this.equals(PdfComment.EOF);
    }
    asString() {
        return new TextDecoder().decode(this.raw);
    }
    isVersionComment() {
        const str = this.asString().toUpperCase();
        return str.startsWith('PDF-');
    }
    isEOFComment() {
        return this.equals(PdfComment.EOF);
    }
    static versionComment(version) {
        const comment = new PdfComment(`PDF-${version}`);
        comment.postTokens = [PdfWhitespaceToken.NEWLINE];
        return comment;
    }
    toJSON() {
        return { type: 'comment', value: this.asString() };
    }
    tokenize() {
        return [new PdfCommentToken(this.raw)];
    }
    cloneImpl() {
        return new PdfComment(new Uint8Array(this.raw));
    }
}
