// pdf-lite/dist/core/objects/pdf-dictionary.js
import { needsCentralWhitespace } from '../../utils/needsCentralWhitespace.js';
import { PdfEndDictionaryToken } from '../tokens/end-dictionary-token.js';
import { PdfStartDictionaryToken } from '../tokens/start-dictionary-token.js';
import { PdfWhitespaceToken } from '../tokens/whitespace-token.js';
import { PdfName } from './pdf-name.js';
import { PdfObject } from './pdf-object.js';
export class PdfDictionary extends PdfObject {
    #entries;
    innerTokens = [];
    constructor(entries) {
        super();
        this.#entries = entries instanceof Map ? entries : new Map();
        if (entries) {
            for (const [key, value] of Object.entries(entries)) {
                this.set(key, value);
            }
        }
    }
    keys() {
        return Array.from(this.#entries.keys());
    }
    get(key) {
        if (key instanceof PdfName) {
            return this.#entries.get(key);
        }
        else {
            for (const [k, v] of this.#entries) {
                if (k.value === key) {
                    return v;
                }
            }
        }
    }
    move(fromKey, toKey) {
        const value = this.get(fromKey);
        if (value !== undefined) {
            this.set(toKey, value);
            this.delete(fromKey);
        }
    }
    set(key, value) {
        const currentValue = this.get(key);
        if (currentValue !== value && !currentValue?.equals(value)) {
            this.modified = true;
        }
        if (key instanceof PdfName) {
            this.#entries.set(key, value);
            return;
        }
        else {
            for (const k of this.#entries.keys()) {
                if (k.value === key) {
                    this.#entries.set(k, value);
                    return;
                }
            }
        }
        this.#entries.set(new PdfName(key), value);
    }
    delete(key) {
        if (this.has(key)) {
            this.modified = true;
        }
        if (key instanceof PdfName) {
            this.#entries.delete(key);
            return;
        }
        else {
            for (const k of this.#entries.keys()) {
                if (k.value === key) {
                    this.#entries.delete(k);
                    return;
                }
            }
        }
    }
    has(key) {
        if (key instanceof PdfName) {
            return this.#entries.has(key);
        }
        else {
            for (const k of this.#entries.keys()) {
                if (k.value === key) {
                    return true;
                }
            }
        }
        return false;
    }
    get values() {
        const obj = {};
        for (const [key, value] of this.#entries) {
            obj[key.value] = value;
        }
        return obj;
    }
    /**
     * Returns an iterator for the dictionary entries.
     * Each entry is a tuple of [key string, value].
     */
    entries() {
        const entries = Array.from(this.#entries.entries()).map(([key, value]) => [key.value, value]);
        return entries[Symbol.iterator]();
    }
    toJSON() {
        const entries = {};
        for (const [key, value] of this.entries()) {
            if (value) {
                entries[key] = value.toJSON();
            }
        }
        return { type: 'dictionary', entries };
    }
    get isTrailingDelimited() {
        return true;
    }
    tokenize() {
        let index = 0;
        const entries = Array.from(this.#entries.entries()).flatMap(([key, value]) => {
            if (value === undefined) {
                return [];
            }
            const preTokens = key.preTokens
                ? []
                : index === 0
                    ? [PdfWhitespaceToken.SPACE]
                    : [];
            const centralTokens = key.postTokens || value.preTokens
                ? []
                : [PdfWhitespaceToken.SPACE];
            const postTokens = value.postTokens
                ? []
                : [PdfWhitespaceToken.SPACE];
            if (needsCentralWhitespace(key, value)) {
                centralTokens.push(PdfWhitespaceToken.SPACE);
            }
            index++;
            return [
                ...preTokens,
                ...key.toTokens(),
                ...centralTokens,
                ...value.toTokens(),
                ...postTokens,
            ];
        });
        return [
            new PdfStartDictionaryToken(),
            ...this.innerTokens,
            ...entries,
            new PdfEndDictionaryToken(),
        ];
    }
    /** Factory-style type conversion: constructs a new instance passing `this` as the first argument */
    becomes(cls, ...args) {
        if (this instanceof cls)
            return this;
        const donor = new cls(this, ...args);
        Object.setPrototypeOf(this, cls.prototype);
        for (const key of Object.getOwnPropertyNames(donor)) {
            if (!Object.prototype.hasOwnProperty.call(this, key)) {
                ;
                this[key] = donor[key];
            }
        }
        return this;
    }
    copyFrom(other) {
        for (const [key, value] of other.#entries) {
            this.#entries.set(key, value);
        }
        this.modified = true;
    }
    cloneImpl() {
        const clonedEntries = new Map();
        for (const [key, value] of this.#entries) {
            clonedEntries.set(key.clone(), value ? value.clone() : undefined);
        }
        const cloned = new PdfDictionary(clonedEntries);
        return cloned;
    }
    setModified(modified) {
        super.setModified(modified);
        for (const value of this.#entries.values()) {
            value?.setModified(modified);
        }
    }
    isModified() {
        return (super.isModified() ||
            Array.from(this.#entries.values()).some((v) => v?.isModified()));
    }
    setImmutable(immutable) {
        super.setImmutable(immutable);
        for (const value of this.#entries.values()) {
            value?.setImmutable(immutable);
        }
    }
}
