// pdf-lite/dist/core/objects/pdf-stream.js
import { PdfEndStreamToken } from '../tokens/end-stream-token.js';
import { PdfStartStreamToken } from '../tokens/start-stream-token.js';
import { PdfStreamChunkToken } from '../tokens/stream-chunk-token.js';
import { PdfDictionary } from './pdf-dictionary.js';
import { PdfObject } from './pdf-object.js';
import { PdfName } from './pdf-name.js';
import { PdfArray } from './pdf-array.js';
import { flate } from '../../filters/flate.js';
import { asciiHex } from '../../filters/asciihex.js';
import { ascii85 } from '../../filters/ascii85.js';
import { lzw } from '../../filters/lzw.js';
import { runLength } from '../../filters/runlength.js';
import { passthroughFilter } from '../../filters/pass-through.js';
import { PdfNumber } from './pdf-number.js';
import { bytesToPdfObjects } from '../generators.js';
import { PdfIndirectObject } from './pdf-indirect-object.js';
import { PdfXRefTableEntry } from './pdf-xref-table.js';
import { stringToBytes } from '../../utils/stringToBytes.js';
import { bytesToString } from '../../utils/bytesToString.js';
export class PdfStreamPredictor extends PdfDictionary {
    static NONE = new PdfStreamPredictor({ Predictor: new PdfNumber(1) });
    static TIFF = new PdfStreamPredictor({ Predictor: new PdfNumber(2) });
    static PNG_NONE = new PdfStreamPredictor({ Predictor: new PdfNumber(10) });
    static PNG_SUB = new PdfStreamPredictor({ Predictor: new PdfNumber(11) });
    static PNG_UP = new PdfStreamPredictor({ Predictor: new PdfNumber(12) });
    static PNG_AVERAGE = new PdfStreamPredictor({
        Predictor: new PdfNumber(13),
    });
    static PNG_PAETH = new PdfStreamPredictor({ Predictor: new PdfNumber(14) });
    static PNG_OPTIMUM = new PdfStreamPredictor({
        Predictor: new PdfNumber(15),
    });
    static XREF_STREAM = new PdfStreamPredictor({
        Predictor: new PdfNumber(12),
        Columns: new PdfNumber(7),
    });
    get predictor() {
        return this.get('Predictor')?.as(PdfNumber)?.value ?? 1;
    }
    set predictor(value) {
        this.set('Predictor', new PdfNumber(value));
    }
    get columns() {
        return this.get('Columns')?.as(PdfNumber)?.value ?? 1;
    }
    set columns(value) {
        this.set('Columns', new PdfNumber(value));
    }
    get colors() {
        return this.get('Colors')?.as(PdfNumber)?.value ?? 1;
    }
    set colors(value) {
        this.set('Colors', new PdfNumber(value));
    }
    get bitsPerComponent() {
        return this.get('BitsPerComponent')?.as(PdfNumber)?.value ?? 8;
    }
    set bitsPerComponent(value) {
        this.set('BitsPerComponent', new PdfNumber(value));
    }
    toDecodeParms() {
        const params = {};
        const predictor = this.get('Predictor')?.as(PdfNumber);
        if (predictor)
            params.Predictor = predictor.value;
        const columns = this.get('Columns')?.as(PdfNumber);
        if (columns)
            params.Columns = columns.value;
        const colors = this.get('Colors')?.as(PdfNumber);
        if (colors)
            params.Colors = colors.value;
        const bpc = this.get('BitsPerComponent')?.as(PdfNumber);
        if (bpc)
            params.BitsPerComponent = bpc.value;
        return params;
    }
    cloneImpl() {
        return PdfStreamPredictor.fromDecodeParms(this.toDecodeParms());
    }
    static fromDecodeParms(params) {
        const predictor = new PdfStreamPredictor();
        if (params.Predictor !== undefined) {
            predictor.predictor = params.Predictor;
        }
        if (params.Columns !== undefined) {
            predictor.columns = params.Columns;
        }
        if (params.Colors !== undefined) {
            predictor.colors = params.Colors;
        }
        if (params.BitsPerComponent !== undefined) {
            predictor.bitsPerComponent = params.BitsPerComponent;
        }
        return predictor;
    }
    static fromDictionary(dict) {
        if (!dict)
            return undefined;
        const predictor = dict.get('Predictor')?.as(PdfNumber)?.value;
        if (predictor === undefined || predictor <= 1 || predictor > 15) {
            return undefined;
        }
        return PdfStreamPredictor.fromDecodeParms({
            Predictor: predictor,
            Columns: dict.get('Columns')?.as(PdfNumber)?.value,
            Colors: dict.get('Colors')?.as(PdfNumber)?.value,
            BitsPerComponent: dict.get('BitsPerComponent')?.as(PdfNumber)
                ?.value,
        });
    }
    encode(data, filterType = 0) {
        const bpp = Math.ceil((this.colors * this.bitsPerComponent) / 8);
        if (this.predictor === 2) {
            return PdfStreamPredictor.tiffEncode(data, this.columns, bpp);
        }
        else if (this.predictor >= 10 && this.predictor <= 15) {
            return PdfStreamPredictor.pngEncode(data, this.columns, bpp, filterType);
        }
        return data;
    }
    decode(data) {
        const bpp = Math.ceil((this.colors * this.bitsPerComponent) / 8);
        if (this.predictor === 2) {
            return PdfStreamPredictor.tiffDecode(data, this.columns, bpp);
        }
        else if (this.predictor >= 10 && this.predictor <= 15) {
            return PdfStreamPredictor.pngDecode(data, this.columns, bpp);
        }
        return data;
    }
    static tiffDecode(data, columns, bpp) {
        const rowLength = columns * bpp;
        const output = new Uint8Array(data.length);
        for (let i = 0; i < data.length; i += rowLength) {
            for (let j = 0; j < rowLength; j++) {
                output[i + j] =
                    j < bpp
                        ? data[i + j]
                        : (data[i + j] + output[i + j - bpp]) & 0xff;
            }
        }
        return output;
    }
    static tiffEncode(data, columns, bpp) {
        const rowLength = columns * bpp;
        const output = new Uint8Array(data.length);
        for (let i = 0; i < data.length; i += rowLength) {
            for (let j = 0; j < rowLength; j++) {
                output[i + j] =
                    j < bpp
                        ? data[i + j]
                        : (data[i + j] - data[i + j - bpp] + 256) & 0xff;
            }
        }
        return output;
    }
    static pngDecode(data, columns, bpp) {
        const rowLength = columns * bpp;
        const output = [];
        for (let i = 0; i < data.length;) {
            const filter = data[i];
            const row = data.slice(i + 1, i + 1 + rowLength);
            const prior = output.slice(output.length - rowLength, output.length);
            const decodedRow = new Uint8Array(rowLength);
            for (let j = 0; j < rowLength; j++) {
                const left = j >= bpp ? decodedRow[j - bpp] : 0;
                const up = prior[j] ?? 0;
                const upperLeft = j >= bpp ? (prior[j - bpp] ?? 0) : 0;
                switch (filter) {
                    case 0:
                        decodedRow[j] = row[j];
                        break;
                    case 1:
                        decodedRow[j] = (row[j] + left) & 0xff;
                        break;
                    case 2:
                        decodedRow[j] = (row[j] + up) & 0xff;
                        break;
                    case 3:
                        decodedRow[j] =
                            (row[j] + Math.floor((left + up) / 2)) & 0xff;
                        break;
                    case 4:
                        decodedRow[j] =
                            (row[j] +
                                this.paethPredictor(left, up, upperLeft)) &
                                0xff;
                        break;
                    default:
                        throw new Error(`Unsupported PNG filter: ${filter}`);
                }
            }
            output.push(...decodedRow);
            i += 1 + rowLength;
        }
        return new Uint8Array(output);
    }
    static pngEncode(data, columns, bpp, filterType) {
        const rowLength = columns * bpp;
        const output = [];
        let prior = new Uint8Array(rowLength);
        for (let i = 0; i < data.length; i += rowLength) {
            const row = data.slice(i, i + rowLength);
            const encodedRow = new Uint8Array(rowLength);
            for (let j = 0; j < rowLength; j++) {
                const left = j >= bpp ? row[j - bpp] : 0;
                const up = prior[j] ?? 0;
                const upperLeft = j >= bpp ? (prior[j - bpp] ?? 0) : 0;
                switch (filterType) {
                    case 0:
                        encodedRow[j] = row[j];
                        break;
                    case 1:
                        encodedRow[j] = (row[j] - left + 256) & 0xff;
                        break;
                    case 2:
                        encodedRow[j] = (row[j] - up + 256) & 0xff;
                        break;
                    case 3:
                        encodedRow[j] =
                            (row[j] - Math.floor((left + up) / 2) + 256) & 0xff;
                        break;
                    case 4:
                        encodedRow[j] =
                            (row[j] -
                                this.paethPredictor(left, up, upperLeft) +
                                256) &
                                0xff;
                        break;
                    default:
                        throw new Error(`Unsupported PNG filter: ${filterType}`);
                }
            }
            output.push(filterType, ...encodedRow);
            prior = row;
        }
        return new Uint8Array(output);
    }
    static paethPredictor(a, b, c) {
        const p = a + b - c;
        const pa = Math.abs(p - a);
        const pb = Math.abs(p - b);
        const pc = Math.abs(p - c);
        if (pa <= pb && pa <= pc)
            return a;
        if (pb <= pc)
            return b;
        return c;
    }
}
export class PdfStream extends PdfObject {
    static PdfStreamPredictor = PdfStreamPredictor;
    header;
    original;
    preStreamDataTokens;
    postStreamDataTokens;
    constructor(options) {
        super();
        if (typeof options === 'string' || options instanceof Uint8Array) {
            options = { header: new PdfDictionary(), original: options };
        }
        this.header = options.header;
        this.modified = options.isModified ?? true;
        this.original =
            typeof options.original === 'string'
                ? stringToBytes(options.original)
                : options.original;
        if (!this.header.get('Length')) {
            this.header.set('Length', new PdfNumber(this.original.length));
        }
    }
    get raw() {
        const length = this.header.get('Length')?.as(PdfNumber)?.value;
        if (length === undefined)
            return this.original;
        return this.original.slice(0, length);
    }
    set raw(data) {
        this.setModified();
        this.original = data;
        this.header.set('Length', new PdfNumber(data.length));
    }
    get originalAsString() {
        return bytesToString(this.original);
    }
    get rawAsString() {
        return bytesToString(this.raw);
    }
    set rawAsString(str) {
        this.raw = stringToBytes(str);
    }
    get data() {
        return this.decode();
    }
    set data(data) {
        const filters = this.getFilters();
        const existingPredictor = this.predictor;
        this.removeAllFilters();
        this.header.delete('DecodeParms');
        this.raw = data;
        if (existingPredictor) {
            this.predictor = existingPredictor;
        }
        for (const filter of filters) {
            this.addFilter(filter);
        }
    }
    get dataAsString() {
        return bytesToString(this.data);
    }
    set dataAsString(str) {
        this.data = stringToBytes(str);
    }
    getFilters() {
        const filters = this.header.get('Filter');
        if (!filters) {
            return [];
        }
        if (filters instanceof PdfName) {
            return [filters.value];
        }
        else if (filters instanceof PdfArray) {
            return filters.items.map((item) => item.value);
        }
        else {
            throw new Error('Invalid Filter entry in PDF stream');
        }
    }
    addFilter(filterName) {
        const filters = this.header.get('Filter');
        if (!filters) {
            this.header.set('Filter', new PdfName(filterName));
        }
        else if (filters instanceof PdfName) {
            this.header.set('Filter', new PdfArray([new PdfName(filterName), filters]));
        }
        else if (filters instanceof PdfArray) {
            filters.items.unshift(new PdfName(filterName));
        }
        else {
            throw new Error('Invalid Filter entry in PDF stream');
        }
        const filter = PdfStream.getFilter(filterName);
        this.raw = filter.encode(this.raw);
        return this;
    }
    get predictor() {
        const raw = this.header.get('DecodeParms');
        if (!raw)
            return undefined;
        if (raw instanceof PdfStreamPredictor) {
            if (raw.get('Predictor') === undefined)
                return undefined;
            return raw;
        }
        if (raw instanceof PdfDictionary) {
            return PdfStreamPredictor.fromDictionary(raw);
        }
        return undefined;
    }
    set predictor(predictor) {
        const cloned = predictor.cloneImpl();
        this.header.set('DecodeParms', cloned);
        this.raw = cloned.encode(this.raw);
    }
    removeFilter(filterName) {
        let filters = this.header.get('Filter');
        if (!filters) {
            return this;
        }
        const decoded = this.decode();
        if (filters instanceof PdfName) {
            if (filters.value === filterName) {
                this.header.delete('Filter');
            }
        }
        else if (filters instanceof PdfArray) {
            filters.items = filters.items.filter((item) => item.value !== filterName);
            if (filters.items.length === 0) {
                this.header.delete('Filter');
            }
            else if (filters.items.length === 1) {
                this.header.set('Filter', filters.items[0]);
            }
        }
        else {
            throw new Error('Invalid Filter entry in PDF stream');
        }
        const finalFilters = this.getFilters();
        const encoded = PdfStream.applyFilters(decoded, finalFilters);
        this.raw = encoded;
        return this;
    }
    removePredictor() {
        const decoded = this.decode();
        this.header.delete('DecodeParms');
        this.raw = decoded;
        return this;
    }
    removeAllFilters() {
        const decoded = this.decode();
        this.raw = decoded;
        this.header.delete('Filter');
        return this;
    }
    decode() {
        let data = this.raw;
        const filters = this.getFilters();
        for (const filterName of filters) {
            const filter = PdfStream.getFilter(filterName);
            data = filter.decode(data);
        }
        const streamPredictor = PdfStreamPredictor.fromDictionary(this.header.get('DecodeParms')?.as(PdfDictionary));
        if (streamPredictor) {
            data = streamPredictor.decode(data);
        }
        return data;
    }
    parseAs(Class) {
        const instance = new Class({
            header: this.header,
            original: this.original,
            isModified: this.isModified(),
        });
        instance.preTokens = this.preTokens;
        instance.postTokens = this.postTokens;
        instance.preStreamDataTokens = this.preStreamDataTokens;
        instance.postStreamDataTokens = this.postStreamDataTokens;
        return instance;
    }
    tokenize() {
        return [
            ...this.header.toTokens(),
            PdfStartStreamToken.withTrailingWhitespace(this.preStreamDataTokens),
            new PdfStreamChunkToken(this.original),
            new PdfEndStreamToken(),
        ];
    }
    toJSON() {
        return {
            type: 'stream',
            header: this.header.toJSON(),
            dataLength: this.raw.length,
        };
    }
    isType(name) {
        const type = this.header.get('Type');
        return type instanceof PdfName && type.value === name;
    }
    static getFilter(name) {
        const allFilters = PdfStream.getAllFilters();
        if (!allFilters[name]) {
            throw new Error(`Unsupported filter: ${name}`);
        }
        return allFilters[name];
    }
    static getAllFilters() {
        return {
            FlateDecode: flate(),
            Fl: flate(),
            ASCIIHexDecode: asciiHex(),
            ASCII85Decode: ascii85(),
            LZWDecode: lzw(),
            RunLengthDecode: runLength(),
            CCITTFaxDecode: passthroughFilter(),
            DCTDecode: passthroughFilter(),
            JPXDecode: passthroughFilter(),
            Crypt: passthroughFilter(),
        };
    }
    static applyFilters(data, filters) {
        let result = data;
        for (const filterName of filters) {
            const filter = PdfStream.getFilter(filterName);
            if (!filter) {
                throw new Error(`Unsupported filter: ${filterName}`);
            }
            result = filter.encode(result);
        }
        return result;
    }
    cloneImpl() {
        return new PdfStream({
            header: this.header.clone(),
            original: new Uint8Array(this.original),
        });
    }
    static fromString(data) {
        return new PdfStream({
            original: stringToBytes(data),
            header: new PdfDictionary(),
        });
    }
}
export class PdfObjStream extends PdfStream {
    _parsedObjects;
    set raw(data) {
        super.raw = data;
        this._parsedObjects = undefined;
    }
    get raw() {
        return super.raw;
    }
    constructor(options) {
        super(options);
        if (!this.isType('ObjStm')) {
            throw new Error('PDF Object Stream must be of type ObjStm');
        }
    }
    static fromObjects(objects) {
        const objByteChunks = [];
        const headerParts = [];
        let offset = 0;
        let objectCount = 0;
        for (const obj of objects) {
            // Use toBytes() to preserve binary content (e.g. UTF-16BE strings)
            // instead of toString() which is Latin-1 and would be re-encoded as UTF-8
            headerParts.push(`${obj.objectNumber} ${offset}`);
            const bytes = obj.content.toBytes();
            objByteChunks.push(bytes, new Uint8Array([0x0a])); // object + newline
            offset += bytes.length + 1;
            objectCount++;
        }
        const headerStr = headerParts.join(' ');
        // First = byte offset of first compressed object from stream start
        const first = objectCount > 0 ? headerStr.length + 1 : 0;
        const headerDict = new PdfDictionary();
        headerDict.set('Type', new PdfName('ObjStm'));
        headerDict.set('N', new PdfNumber(objectCount));
        headerDict.set('First', new PdfNumber(first));
        if (objectCount === 0) {
            return new PdfObjStream({
                header: headerDict,
                original: new Uint8Array(),
            });
        }
        // Build stream bytes: headerStr + '\n' + obj1bytes + obj2bytes + ...
        // Drop trailing newline after the last object
        objByteChunks.pop();
        const headerBytes = stringToBytes(headerStr + '\n');
        const totalLength = objByteChunks.reduce((sum, chunk) => sum + chunk.length, headerBytes.length);
        const streamBytes = new Uint8Array(totalLength);
        let pos = 0;
        streamBytes.set(headerBytes, pos);
        pos += headerBytes.length;
        for (const chunk of objByteChunks) {
            streamBytes.set(chunk, pos);
            pos += chunk.length;
        }
        return new PdfObjStream({
            header: headerDict,
            original: streamBytes,
        });
    }
    *getObjectStream() {
        const decodedData = this.decode();
        const reader = bytesToPdfObjects([decodedData]);
        const numbers = [];
        let i = 0;
        const n = this.header.get('N');
        const totalObjects = n ? n.value : 0;
        while (true) {
            const { value: obj, done } = reader.next();
            if (done)
                break;
            if (obj instanceof PdfNumber &&
                (totalObjects === 0 || numbers.length < totalObjects * 2)) {
                // Collect object numbers and byte offsets from the header section.
                // Stop once we have all 2*N pairs so that object values that happen
                // to be plain integers are not mistakenly treated as header entries.
                numbers.push(obj);
            }
            else {
                // This is an actual PDF object (can be Dictionary, Array, String, Name, etc.)
                // The first N*2 numbers are: obj_num1 offset1 obj_num2 offset2 ...
                // After that come the actual objects
                const objectNumber = numbers[i * 2].value;
                const generationNumber = 0;
                const decompressedObject = new PdfIndirectObject({
                    objectNumber,
                    generationNumber,
                    content: obj,
                    compressed: true,
                }).clone();
                decompressedObject.setModified(false);
                yield decompressedObject;
                i++;
                // Stop after we've read N objects
                if (totalObjects > 0 && i >= totalObjects) {
                    break;
                }
            }
        }
    }
    getParsedObjects() {
        if (!this._parsedObjects) {
            this._parsedObjects = new Map();
            for (const obj of this.getObjectStream()) {
                this._parsedObjects.set(obj.objectNumber, obj);
            }
        }
        return this._parsedObjects;
    }
    getObject(options) {
        return this.getParsedObjects().get(options.objectNumber);
    }
    getObjects() {
        return Array.from(this.getParsedObjects().values());
    }
    cloneImpl() {
        return new PdfObjStream({
            header: this.header.clone(),
            original: new Uint8Array(this.original),
        });
    }
}
export class PdfXRefStreamCompressedEntry {
    objectNumber;
    objectStreamNumber;
    index;
    constructor(options) {
        this.objectNumber =
            options.objectNumber instanceof PdfNumber
                ? options.objectNumber
                : new PdfNumber(options.objectNumber);
        this.objectStreamNumber =
            options.objectStreamNumber instanceof PdfNumber
                ? options.objectStreamNumber
                : new PdfNumber(options.objectStreamNumber);
        this.index =
            options.index instanceof PdfNumber
                ? options.index
                : new PdfNumber(options.index);
    }
}
export const PdfXRefStreamStandardEntry = PdfXRefTableEntry;
export class PdfXRefStream extends PdfStream {
    constructor(options) {
        super({
            header: options?.header ?? PdfXRefStream.createNewHeader(),
            original: options?.original ?? new Uint8Array(),
            isModified: options?.isModified,
        });
        if (!this.isType('XRef')) {
            throw new Error('PDF XRef Stream must be of type XRef');
        }
    }
    static createNewHeader() {
        const headerDict = new PdfDictionary();
        headerDict.set('Type', new PdfName('XRef'));
        headerDict.set('W', new PdfArray([
            new PdfNumber(1),
            new PdfNumber(4),
            new PdfNumber(2),
        ]));
        headerDict.set('Size', new PdfNumber(0));
        return headerDict;
    }
    static fromEntries(entries, headerDict = new PdfDictionary()) {
        headerDict.delete('DecodeParms');
        headerDict.delete('Filter');
        headerDict.set('Type', new PdfName('XRef'));
        entries.sort((a, b) => a.objectNumber.value - b.objectNumber.value);
        const W = this.calculateW(entries);
        headerDict.set('W', new PdfArray(W.map((w) => new PdfNumber(w))));
        // Build Index array - pairs of [start, count] for each contiguous range
        const indexArray = [];
        if (entries.length > 0) {
            let rangeStart = entries[0].objectNumber.value;
            let rangeCount = 1;
            for (let i = 1; i < entries.length; i++) {
                const currentNum = entries[i].objectNumber.value;
                const prevNum = entries[i - 1].objectNumber.value;
                if (currentNum === prevNum + 1) {
                    // Contiguous - extend current range
                    rangeCount++;
                }
                else {
                    // Gap found - save current range and start new one
                    indexArray.push(rangeStart, rangeCount);
                    rangeStart = currentNum;
                    rangeCount = 1;
                }
            }
            // Add final range
            indexArray.push(rangeStart, rangeCount);
        }
        headerDict.set('Index', new PdfArray(indexArray.map((num) => new PdfNumber(num))));
        const size = entries.length === 0
            ? 0
            : Math.max(...entries.map((entry) => entry.objectNumber.value)) + 1;
        headerDict.set('Size', new PdfNumber(size));
        // Encode entries to bytes
        const entrySize = W.reduce((a, b) => a + b, 0);
        const streamBytes = new Uint8Array(entries.length * entrySize);
        let offset = 0;
        for (const entry of entries) {
            const typeByte = entry instanceof PdfXRefTableEntry ? (entry.inUse ? 1 : 0) : 2;
            let field1 = typeByte;
            let field2 = 0;
            let field3 = 0;
            if (entry instanceof PdfXRefTableEntry) {
                field2 = entry.byteOffset.value;
                field3 = entry.generationNumber.value;
            }
            else if (entry instanceof PdfXRefStreamCompressedEntry) {
                field2 = entry.objectStreamNumber.value;
                field3 = entry.index.value;
            }
            for (let i = 0; i < W[0]; i++) {
                streamBytes[offset++] = (field1 >> (8 * (W[0] - 1 - i))) & 0xff;
            }
            for (let i = 0; i < W[1]; i++) {
                streamBytes[offset++] = (field2 >> (8 * (W[1] - 1 - i))) & 0xff;
            }
            for (let i = 0; i < W[2]; i++) {
                streamBytes[offset++] = (field3 >> (8 * (W[2] - 1 - i))) & 0xff;
            }
        }
        const rawData = new Uint8Array(streamBytes);
        headerDict.set('Length', new PdfNumber(rawData.length));
        return new PdfXRefStream({
            header: headerDict,
            original: rawData,
        });
    }
    addEntry(entry) {
        const entries = this.getEntries();
        entries.push(entry);
        const newXRefStream = PdfXRefStream.fromEntries(entries);
        this.header = newXRefStream.header;
        this.raw = newXRefStream.raw;
        return this;
    }
    static calculateW(entries) {
        let maxGeneration = 0;
        const maxOffset = Math.max(...entries.map((entry) => {
            let value = 0;
            if (entry instanceof PdfXRefTableEntry) {
                value = entry.byteOffset.value;
                maxGeneration = Math.max(maxGeneration, entry.generationNumber.value);
            }
            else if (entry instanceof PdfXRefStreamCompressedEntry) {
                value = entry.objectStreamNumber.value;
            }
            return value;
        }));
        const W1 = 1; // Type byte
        const W2 = maxOffset === -1 ? 0 : Math.ceil(Math.log2(maxOffset + 1) / 8) || 1; // Offset or next free
        const W3 = Math.ceil(Math.log2(maxGeneration + 1) / 8) || 1; // Generation number or index
        return [W1, W2, W3];
    }
    get prev() {
        const prev = this.header.get('Prev');
        return prev instanceof PdfNumber ? prev : undefined;
    }
    *getEntryStream() {
        const data = this.decode();
        const header = this.header;
        const W = header
            .get('W')
            ?.as((PdfArray))
            ?.items?.map((v) => v.value);
        if (!W)
            throw new Error('Missing W entry in XRef stream');
        if (W.length !== 3) {
            throw new Error('Invalid W entry in XRef stream. Expected array of 3 numbers.');
        }
        const entrySize = W[0] + W[1] + W[2];
        const Index = header
            .get('Index')
            ?.as((PdfArray))
            ?.items?.map((v) => v.value);
        const Size = header.get('Size').as(PdfNumber).value;
        const objectRanges = Index
            ? Array.from({ length: Index.length / 2 }, (_, i) => [
                Index[i * 2],
                Index[i * 2 + 1],
            ])
            : [[0, Size]];
        const expectedLength = objectRanges.reduce((sum, [, count]) => sum + count, 0) * entrySize;
        if (data.length < expectedLength) {
            throw new Error(`XRef stream too short: expected ${expectedLength} bytes, got ${data.length}`);
        }
        let offset = 0;
        for (const [startObj, count] of objectRanges) {
            for (let i = 0; i < count; i++) {
                const objectNumber = startObj + i;
                const type = this.readInt(data, offset, W[0]);
                offset += W[0];
                const field2 = this.readInt(data, offset, W[1]);
                offset += W[1];
                const field3 = this.readInt(data, offset, W[2]);
                offset += W[2];
                let entry;
                if (type === 0) {
                    entry = new PdfXRefTableEntry({
                        objectNumber,
                        generationNumber: field3,
                        byteOffset: field2,
                        inUse: false,
                    });
                }
                else if (type === 1) {
                    entry = new PdfXRefTableEntry({
                        objectNumber,
                        generationNumber: field3,
                        byteOffset: field2,
                        inUse: true,
                    });
                }
                else if (type === 2) {
                    entry = new PdfXRefStreamCompressedEntry({
                        objectNumber,
                        objectStreamNumber: field2,
                        index: field3,
                    });
                }
                else {
                    throw new Error(`Unknown xref entry type: ${type}`);
                }
                yield entry;
            }
        }
    }
    getEntries() {
        return Array.from(this.getEntryStream());
    }
    readInt(data, offset, length) {
        let value = 0;
        for (let i = 0; i < length; i++) {
            value = (value << 8) | data[offset + i];
        }
        return value;
    }
    cloneImpl() {
        return new PdfXRefStream({
            header: this.header.clone(),
            original: new Uint8Array(this.original),
        });
    }
}
