// pdf-lite/dist/core/objects/pdf-start-xref.js
import { PdfStartXRefToken } from '../tokens/start-xref-token.js';
import { PdfWhitespaceToken } from '../tokens/whitespace-token.js';
import { PdfNumber } from './pdf-number.js';
import { PdfObject } from './pdf-object.js';
export class PdfStartXRef extends PdfObject {
    offset;
    constructor(offset = -1) {
        // default to -1 for unknown offset
        super();
        this.offset =
            offset instanceof PdfNumber
                ? offset
                : new PdfNumber({ value: offset });
        this.offset.isByteOffset = true;
    }
    toJSON() {
        return { type: 'start-xref', offset: this.offset.value };
    }
    tokenize() {
        const whiteSpaceTokens = this.offset.preTokens
            ? []
            : [PdfWhitespaceToken.NEWLINE];
        return [
            new PdfStartXRefToken(),
            ...whiteSpaceTokens,
            ...this.offset.toTokens(),
        ];
    }
    cloneImpl() {
        const cloned = new PdfStartXRef(this.offset.clone());
        return cloned;
    }
    setModified(modified) {
        super.setModified(modified);
        this.offset.setModified(modified);
    }
    isModified() {
        return super.isModified() || this.offset.isModified();
    }
}
