// pdf-lite/dist/core/objects/pdf-object-reference.js
import { PdfObjectReferenceToken } from '../tokens/object-reference-token.js';
import { PdfObject } from './pdf-object.js';
export class PdfObjectReference extends PdfObject {
    objectNumber;
    generationNumber;
    resolver;
    constructor(objectNumber, generationNumber) {
        super();
        this.objectNumber = objectNumber;
        this.generationNumber = generationNumber;
    }
    tokenize() {
        return [
            new PdfObjectReferenceToken(this.objectNumber, this.generationNumber),
        ];
    }
    cloneImpl() {
        const cloned = new PdfObjectReference(this.objectNumber, this.generationNumber);
        return cloned;
    }
    toJSON() {
        return {
            type: 'reference',
            objectNumber: this.objectNumber,
            generationNumber: this.generationNumber,
        };
    }
    resolve(cls) {
        if (!this.resolver) {
            throw new Error(`No resolver set for PdfObjectReference '${this.objectNumber} ${this.generationNumber}'`);
        }
        const object = this.resolver.resolve(this.objectNumber, this.generationNumber);
        if (cls) {
            return object.becomes(cls);
        }
        return object;
    }
    get key() {
        return `${this.objectNumber}/${this.generationNumber}`;
    }
}
