// pdf-lite/dist/annotations/pdf-annotation.js
import { PdfDictionary } from '../core/objects/pdf-dictionary.js';
import { PdfIndirectObject } from '../core/objects/pdf-indirect-object.js';
import { PdfObjectReference } from '../core/objects/pdf-object-reference.js';
import { PdfArray } from '../core/objects/pdf-array.js';
import { PdfNumber } from '../core/objects/pdf-number.js';
import { PdfAnnotationFlags } from './pdf-annotation-flags.js';
import { PdfPage } from '../pdf/pdf-page.js';
/**
 * Base class for all PDF annotations.
 */
export class PdfAnnotation extends PdfIndirectObject {
    _annotationFlags;
    get flags_() {
        if (!this._annotationFlags) {
            const flagValue = this.content.get('F')?.as(PdfNumber)?.value ?? 0;
            this._annotationFlags = new PdfAnnotationFlags(flagValue);
            this.content.set('F', this._annotationFlags);
        }
        return this._annotationFlags;
    }
    constructor(options) {
        super(options?.other ??
            new PdfIndirectObject({ content: new PdfDictionary() }));
        // Eagerly initialize when constructed normally
        this.flags_;
    }
    get rect() {
        const rectArray = this.content.get('Rect')?.as((PdfArray));
        if (!rectArray)
            return null;
        return rectArray.items.map((num) => num.value);
    }
    set rect(rect) {
        if (rect === null) {
            this.content.delete('Rect');
            return;
        }
        const rectArray = new PdfArray(rect.map((num) => new PdfNumber(num)));
        this.content.set('Rect', rectArray);
    }
    get annotationFlags() {
        return this.flags_.annotationFlags;
    }
    set annotationFlags(flags) {
        this.flags_.annotationFlags = flags;
    }
    get invisible() {
        return this.flags_.invisible;
    }
    set invisible(value) {
        this.flags_.invisible = value;
    }
    get hidden() {
        return this.flags_.hidden;
    }
    set hidden(value) {
        this.flags_.hidden = value;
    }
    get print() {
        return this.flags_.print;
    }
    set print(value) {
        this.flags_.print = value;
    }
    get noZoom() {
        return this.flags_.noZoom;
    }
    set noZoom(value) {
        this.flags_.noZoom = value;
    }
    get noRotate() {
        return this.flags_.noRotate;
    }
    set noRotate(value) {
        this.flags_.noRotate = value;
    }
    get noView() {
        return this.flags_.noView;
    }
    set noView(value) {
        this.flags_.noView = value;
    }
    get locked() {
        return this.flags_.locked;
    }
    set locked(value) {
        this.flags_.locked = value;
    }
    get appearanceStreamDict() {
        const apDict = this.content.get('AP');
        if (!apDict)
            return null;
        if (apDict instanceof PdfObjectReference) {
            const resolved = apDict.resolve();
            if (resolved.content instanceof PdfDictionary) {
                return resolved.content;
            }
            else {
                throw new Error('Invalid AP reference: not a dictionary');
            }
        }
        else if (apDict instanceof PdfDictionary) {
            return apDict;
        }
        else {
            throw new Error('Invalid AP entry: must be a dictionary or reference');
        }
    }
    set appearanceStreamDict(dict) {
        if (dict === null) {
            this.content.delete('AP');
            return;
        }
        this.content.set('AP', dict);
    }
    get parentRef() {
        const ref = this.content.get('P')?.as(PdfObjectReference);
        return ref ?? null;
    }
    set parentRef(ref) {
        if (ref === null) {
            this.content.delete('P');
        }
        else {
            this.content.set('P', ref);
        }
    }
    get page() {
        return this.parentRef?.resolve(PdfPage) ?? null;
    }
}
