// pdf-lite/dist/annotations/pdf-annotation-flags.js
import { PdfNumber } from '../core/objects/pdf-number.js';
/**
 * Provides annotation flag (F field) accessors for PDF annotations.
 * These are generic to all annotation types per the PDF spec.
 */
export class PdfAnnotationFlags extends PdfNumber {
    constructor(value = 0) {
        super(value);
    }
    get annotationFlags() {
        return this.value;
    }
    set annotationFlags(flags) {
        this.value = flags;
    }
    get invisible() {
        return (this.annotationFlags & 1) !== 0;
    }
    set invisible(value) {
        if (value) {
            this.annotationFlags = this.annotationFlags | 1;
        }
        else {
            this.annotationFlags = this.annotationFlags & ~1;
        }
    }
    get hidden() {
        return (this.annotationFlags & 2) !== 0;
    }
    set hidden(value) {
        if (value) {
            this.annotationFlags = this.annotationFlags | 2;
        }
        else {
            this.annotationFlags = this.annotationFlags & ~2;
        }
    }
    get print() {
        return (this.annotationFlags & 4) !== 0;
    }
    set print(value) {
        if (value) {
            this.annotationFlags = this.annotationFlags | 4;
        }
        else {
            this.annotationFlags = this.annotationFlags & ~4;
        }
    }
    get noZoom() {
        return (this.annotationFlags & 8) !== 0;
    }
    set noZoom(value) {
        if (value) {
            this.annotationFlags = this.annotationFlags | 8;
        }
        else {
            this.annotationFlags = this.annotationFlags & ~8;
        }
    }
    get noRotate() {
        return (this.annotationFlags & 16) !== 0;
    }
    set noRotate(value) {
        if (value) {
            this.annotationFlags = this.annotationFlags | 16;
        }
        else {
            this.annotationFlags = this.annotationFlags & ~16;
        }
    }
    get noView() {
        return (this.annotationFlags & 32) !== 0;
    }
    set noView(value) {
        if (value) {
            this.annotationFlags = this.annotationFlags | 32;
        }
        else {
            this.annotationFlags = this.annotationFlags & ~32;
        }
    }
    get locked() {
        return (this.annotationFlags & 128) !== 0;
    }
    set locked(value) {
        if (value) {
            this.annotationFlags = this.annotationFlags | 128;
        }
        else {
            this.annotationFlags = this.annotationFlags & ~128;
        }
    }
}
