// pdf-lite/dist/acroform/js/pdf-field-actions.js
import { PdfDictionary } from '../../core/objects/pdf-dictionary.js';
import { PdfObjectReference } from '../../core/objects/pdf-object-reference.js';
import { PdfJavaScriptAction } from './pdf-javascript-action.js';
export class PdfFieldActions extends PdfDictionary {
    activateDict;
    engine;
    constructor(dict, options) {
        super();
        this.copyFrom(dict);
        this.activateDict = options?.activateDict;
        this.engine = options?.engine;
    }
    _resolve(key) {
        const entry = this.get(key);
        if (!entry)
            return null;
        let actionDict;
        if (entry instanceof PdfObjectReference) {
            const resolved = entry.resolve();
            if (resolved?.content instanceof PdfDictionary) {
                actionDict = resolved.content;
            }
        }
        else if (entry instanceof PdfDictionary) {
            actionDict = entry;
        }
        if (!actionDict)
            return null;
        return actionDict.becomes(PdfJavaScriptAction, {
            engine: this.engine,
        });
    }
    get keystroke() {
        return this._resolve('K');
    }
    get validate() {
        return this._resolve('V');
    }
    get calculate() {
        return this._resolve('C');
    }
    get format() {
        return this._resolve('F');
    }
    get activate() {
        if (!this.activateDict)
            return null;
        return this.activateDict.becomes(PdfJavaScriptAction, {
            engine: this.engine,
        });
    }
}
