// pki-lite/dist/x509/attributes/RevocationValues.js
import { Any } from '../../asn1/Any.js';
import { ObjectIdentifier } from '../../asn1/ObjectIdentifier.js';
import { asn1js, PkiBase, } from '../../core/PkiBase.js';
import { BasicOCSPResponse } from '../../ocsp/BasicOCSPResponse.js';
import { CertificateList } from '../CertificateList.js';
import { Asn1ParseError } from '../../core/errors/Asn1ParseError.js';
/**
 * Represents the OtherRevVals structure used in .
 *
 * @asn
 * ```asn
 * OtherRevVals ::= SEQUENCE {
 *     OtherRevValType   OtherRevValType,
 *     OtherRevVals      ANY DEFINED BY OtherRevValType
 * }
 * ```
 */
export class OtherRevVals extends PkiBase {
    OtherRevValType;
    OtherRevVals;
    constructor(options) {
        super();
        this.OtherRevValType = new ObjectIdentifier({
            value: options.OtherRevValType,
        });
        this.OtherRevVals = new Any({ derBytes: options.OtherRevVals });
    }
    toAsn1() {
        return new asn1js.Sequence({
            value: [this.OtherRevValType.toAsn1(), this.OtherRevVals.toAsn1()],
        });
    }
    static fromAsn1(asn1) {
        if (!(asn1 instanceof asn1js.Sequence)) {
            throw new Asn1ParseError('Invalid ASN.1 structure: expected SEQUENCE');
        }
        const type = asn1.valueBlock.value[0];
        const value = asn1.valueBlock.value[1];
        return new OtherRevVals({
            OtherRevValType: type.valueBlock.toString(),
            OtherRevVals: new Any({ derBytes: value }),
        });
    }
}
/**
 * Represents the RevocationValues attribute defined by https://datatracker.ietf.org/doc/html/rfc5126#page-54.
 *
 * @asn
 * ```asn
 * RevocationValues ::=  SEQUENCE {
 *    crlVals          [0] SEQUENCE OF CertificateList OPTIONAL,
 *    ocspVals         [1] SEQUENCE OF BasicOCSPResponse OPTIONAL,
 *    otherRevVals     [2] OtherRevVals OPTIONAL
 * }
 * ```
 */
export class RevocationValues extends PkiBase {
    crlVals;
    ocspVals;
    otherRevVals;
    constructor(options) {
        super();
        this.crlVals = options.crlVals;
        this.ocspVals = options.ocspVals;
        this.otherRevVals = options.otherRevVals;
    }
    toAsn1() {
        const values = [];
        if (this.crlVals && this.crlVals.length > 0) {
            const crlSeq = new asn1js.Sequence({
                value: this.crlVals.map((crl) => crl.toAsn1()),
            });
            values.push(new asn1js.Constructed({
                idBlock: { tagClass: 3, tagNumber: 0 },
                value: [crlSeq],
            }));
        }
        if (this.ocspVals && this.ocspVals.length > 0) {
            const ocspSeq = new asn1js.Sequence({
                value: this.ocspVals.map((ocsp) => ocsp.toAsn1()),
            });
            values.push(new asn1js.Constructed({
                idBlock: { tagClass: 3, tagNumber: 1 },
                value: [ocspSeq],
            }));
        }
        if (this.otherRevVals && this.otherRevVals.length > 0) {
            const otherRevValsSeq = new asn1js.Sequence({
                value: this.otherRevVals.map((ori) => ori.toAsn1()),
            });
            values.push(new asn1js.Constructed({
                idBlock: { tagClass: 3, tagNumber: 2 },
                value: [otherRevValsSeq],
            }));
        }
        return new asn1js.Sequence({ value: values });
    }
    static fromAsn1(asn1) {
        if (!(asn1 instanceof asn1js.Sequence)) {
            throw new Asn1ParseError('Invalid ASN.1 structure: expected SEQUENCE');
        }
        const crls = [];
        const ocsps = [];
        const otherRevVals = [];
        for (const element of asn1.valueBlock.value) {
            if (!(element instanceof asn1js.Constructed)) {
                continue;
            }
            const tagNumber = element.idBlock.tagNumber;
            if (tagNumber === 0) {
                const crlSeq = element.valueBlock.value[0];
                for (const crlAsn1 of crlSeq.valueBlock.value) {
                    crls.push(CertificateList.fromAsn1(crlAsn1));
                }
            }
            else if (tagNumber === 1) {
                const ocspSeq = element.valueBlock.value[0];
                for (const ocspAsn1 of ocspSeq.valueBlock.value) {
                    ocsps.push(BasicOCSPResponse.fromAsn1(ocspAsn1));
                }
            }
            else if (tagNumber === 2) {
                const oriSeq = element.valueBlock.value[0];
                for (const oriAsn1 of oriSeq.valueBlock.value) {
                    otherRevVals.push(OtherRevVals.fromAsn1(oriAsn1));
                }
            }
        }
        return new RevocationValues({
            crlVals: crls,
            ocspVals: ocsps,
            otherRevVals,
        });
    }
}
