// pdf-lite/dist/acroform/xfa/pdf-xfa-data.js
import { PdfIndirectObject } from '../../core/objects/pdf-indirect-object.js';
import { Xml } from '../../utils/xml.js';
/**
 * Wraps an XFA datasets stream as a typed PDF indirect object.
 * Provides methods to read/write XML and update individual field values.
 */
export class PdfXfaData extends PdfIndirectObject {
    constructor(stream) {
        super(stream);
    }
    get xml() {
        return this.content.dataAsString;
    }
    set xml(xml) {
        this.content.dataAsString = xml;
    }
    updateField(name, value) {
        this.xml = PdfXfaData.updateFieldValue(this.xml, name, value);
    }
    updateFields(fields) {
        if (fields.length === 0)
            return;
        let xml = this.xml;
        for (const { name, value } of fields) {
            xml = PdfXfaData.updateFieldValue(xml, name, value);
        }
        this.xml = xml;
    }
    getFieldValue(name) {
        const leafName = PdfXfaData.leafName(name);
        if (!leafName)
            return null;
        return Xml.getElementContent(this.xml, leafName);
    }
    importData(fields) {
        for (const field in fields) {
            const value = fields[field];
            if (value === undefined)
                continue;
            this.updateField(field, value);
        }
    }
    static updateFieldValue(xml, fieldName, value) {
        const leaf = PdfXfaData.leafName(fieldName);
        if (!leaf)
            return xml;
        const escapedValue = Xml.escapeValue(value);
        // Try updating an existing element
        const updated = Xml.setElementContent(xml, leaf, escapedValue);
        if (updated !== xml)
            return updated;
        // Element doesn't exist — create the missing hierarchy
        const segments = PdfXfaData.cleanSegments(fieldName);
        if (segments.length === 0)
            return xml;
        // Find the deepest existing ancestor (skip the leaf itself)
        let insertAt = -1;
        for (let i = segments.length - 2; i >= 0; i--) {
            if (Xml.hasElement(xml, segments[i])) {
                insertAt = i;
                break;
            }
        }
        if (insertAt === -1)
            return xml;
        const childXml = Xml.wrapInElements(escapedValue, segments.slice(insertAt + 1));
        return Xml.insertChild(xml, segments[insertAt], childXml);
    }
    static leafName(fieldName) {
        const segments = fieldName.split('.');
        const leaf = segments[segments.length - 1].replace(/\[\d+\]$/, '');
        return leaf.startsWith('#') ? null : leaf;
    }
    static cleanSegments(fieldName) {
        return fieldName
            .split('.')
            .map((s) => s.replace(/\[\d+\]$/, ''))
            .filter((s) => !s.startsWith('#'));
    }
}
