// pdf-lite/dist/pdf/pdf-page.js
import { PdfDictionary } from '../core/objects/pdf-dictionary.js';
import { PdfArray } from '../core/objects/pdf-array.js';
import { PdfObjectReference } from '../core/objects/pdf-object-reference.js';
import { PdfIndirectObject } from '../core/objects/pdf-indirect-object.js';
import { PdfNumber } from '../core/objects/pdf-number.js';
import { PdfPages } from './pdf-pages.js';
export class PdfPage extends PdfIndirectObject {
    constructor(options) {
        super(options ??
            new PdfIndirectObject({
                content: new PdfDictionary(),
            }));
    }
    getBox(name) {
        const box = this.content.get(name)?.as((PdfArray));
        if (!box || box.items.length < 4)
            return null;
        return [
            box.items[0].value,
            box.items[1].value,
            box.items[2].value,
            box.items[3].value,
        ];
    }
    setBox(name, value) {
        if (value === null) {
            this.content.delete(name);
        }
        else {
            this.content.set(name, new PdfArray(value.map((v) => new PdfNumber(v))));
        }
    }
    get mediaBox() {
        return this.getBox('MediaBox') ?? [0, 0, 612, 792];
    }
    set mediaBox(value) {
        this.setBox('MediaBox', value);
    }
    get cropBox() {
        return this.getBox('CropBox');
    }
    set cropBox(value) {
        this.setBox('CropBox', value);
    }
    get bleedBox() {
        return this.getBox('BleedBox');
    }
    set bleedBox(value) {
        this.setBox('BleedBox', value);
    }
    get trimBox() {
        return this.getBox('TrimBox');
    }
    set trimBox(value) {
        this.setBox('TrimBox', value);
    }
    get artBox() {
        return this.getBox('ArtBox');
    }
    set artBox(value) {
        this.setBox('ArtBox', value);
    }
    get width() {
        const box = this.mediaBox;
        return box[2] - box[0];
    }
    get height() {
        const box = this.mediaBox;
        return box[3] - box[1];
    }
    get rotate() {
        return this.content.get('Rotate')?.as(PdfNumber)?.value ?? 0;
    }
    set rotate(value) {
        this.content.set('Rotate', new PdfNumber(value));
    }
    get contents() {
        const entry = this.content.get('Contents');
        if (!entry)
            return null;
        if (entry instanceof PdfArray)
            return entry;
        return entry.as(PdfObjectReference) ?? null;
    }
    set contents(value) {
        if (value === null) {
            this.content.delete('Contents');
        }
        else {
            this.content.set('Contents', value);
        }
    }
    get resources() {
        const entry = this.content.get('Resources');
        if (!entry)
            return null;
        if (entry instanceof PdfObjectReference) {
            const resolved = entry.resolve();
            if (resolved?.content instanceof PdfDictionary) {
                return resolved.content;
            }
            return null;
        }
        return entry.as(PdfDictionary) ?? null;
    }
    set resources(value) {
        if (value === null) {
            this.content.delete('Resources');
        }
        else {
            this.content.set('Resources', value);
        }
    }
    get annotations() {
        const entry = this.content.get('Annots');
        if (entry instanceof PdfArray) {
            return entry;
        }
        if (entry instanceof PdfObjectReference) {
            const resolved = entry.resolve();
            if (resolved?.content instanceof PdfArray) {
                return resolved.content;
            }
        }
        const newArray = new PdfArray();
        this.content.set('Annots', newArray);
        return newArray;
    }
    set annotations(value) {
        this.content.set('Annots', value);
    }
    get parentRef() {
        return this.content.get('Parent').as(PdfObjectReference);
    }
    set parentRef(value) {
        this.content.set('Parent', value);
    }
    get parent() {
        return this.parentRef.resolve(PdfPages);
    }
    set parent(value) {
        this.parentRef = value.reference;
    }
}
