// pdf-lite/dist/pdf/pdf-document.js
import { PdfObject } from '../core/objects/pdf-object.js';
import { PdfStandardSecurityHandler, } from '../security/handlers/base.js';
import { createFromDictionary } from '../security/handlers/utils.js';
import { PdfIndirectObject } from '../core/objects/pdf-indirect-object.js';
import { PdfComment } from '../core/objects/pdf-comment.js';
import { PdfWhitespaceToken } from '../core/tokens/whitespace-token.js';
import { PdfObjStream, PdfStream, PdfXRefStreamCompressedEntry, } from '../core/objects/pdf-stream.js';
import { PdfArray } from '../core/objects/pdf-array.js';
import { PdfDictionary } from '../core/objects/pdf-dictionary.js';
import { PdfObjectReference, } from '../core/objects/pdf-object-reference.js';
import { PdfTokenSerializer } from '../core/serializer.js';
import { PdfRevision } from './pdf-revision.js';
import { PdfV5SecurityHandler } from '../security/handlers/v5.js';
import { PdfByteOffsetToken } from '../core/tokens/byte-offset-token.js';
import { PdfNumberToken } from '../core/tokens/number-token.js';
import { PdfXRefTableEntryToken } from '../core/tokens/xref-table-entry-token.js';
import { PdfStartXRef } from '../core/objects/pdf-start-xref.js';
import { FoundCompressedObjectError } from '../errors.js';
import { PdfReader } from './pdf-reader.js';
import { PdfSigner } from '../signing/signer.js';
import { concatUint8Arrays } from '../utils/concatUint8Arrays.js';
import { PdfAcroForm } from '../acroform/pdf-acro-form.js';
import { PdfPages } from './pdf-pages.js';
/**
 * Represents a PDF document with support for reading, writing, and modifying PDF files.
 * Handles document structure, revisions, encryption, and digital signatures.
 *
 * @example
 * ```typescript
 * // Create a new document
 * const document = new PdfDocument()
 *
 * // Read from bytes
 * const document = await PdfDocument.fromBytes(fileBytes)
 *
 * // Add objects and commit
 * document.add(pdfObject)
 * await document.commit()
 * ```
 */
export class PdfDocument extends PdfObject {
    /** List of document revisions for incremental updates */
    revisions;
    /** Signer instance for digital signature operations */
    signer;
    /** Security handler for encryption/decryption operations */
    securityHandler;
    /** Whether the document is currently in incremental mode (appending changes as a new revision) */
    incremental = false;
    originalSecurityHandler;
    hasEncryptionDictionary = false;
    _resolvedCache = new Map();
    _objStreamCache = new Map();
    _committing = false;
    _updating = false;
    _finalized = false;
    _signed = false;
    /**
     * Creates a new PDF document instance.
     *
     * @param options - Configuration options for the document
     * @param options.revisions - Pre-existing revisions for the document
     * @param options.version - PDF version string (e.g., '1.7', '2.0') or version comment
     * @param options.password - User password for encryption
     * @param options.ownerPassword - Owner password for encryption
     * @param options.securityHandler - Custom security handler for encryption
     * @param options.signer - Custom signer for digital signatures
     */
    constructor(options) {
        super();
        this.revisions = options?.revisions ?? [new PdfRevision()];
        if (options?.version instanceof PdfComment) {
            this.header = options.version;
        }
        else {
            this.setVersion(options?.version ?? '2.0');
        }
        if (options?.password) {
            this.setPassword(options.password);
        }
        if (options?.ownerPassword) {
            this.setOwnerPassword(options.ownerPassword);
        }
        this.signer = options?.signer ?? new PdfSigner({ document: this });
        this.linkRevisions();
        this.wireResolvers(...this.objects.filter((x) => x instanceof PdfIndirectObject), ...this.revisions.map((rev) => rev.xref.trailerDict));
        this.calculateOffsets();
        this.originalSecurityHandler = options?.securityHandler;
        this.resetSecurityHandler();
    }
    resolve(objectNumber, generationNumber) {
        const cacheKey = `${objectNumber} ${generationNumber}`;
        const cached = this._resolvedCache.get(cacheKey);
        if (cached)
            return cached;
        const found = this.readObject({ objectNumber, generationNumber });
        if (!found) {
            throw new Error(`Object ${objectNumber} ${generationNumber} not found`);
        }
        this.wireResolvers(found);
        this._resolvedCache.set(cacheKey, found);
        return found;
    }
    get acroform() {
        const root = this.root;
        const acroFormEntry = root.content.get('AcroForm');
        if (!acroFormEntry)
            return null;
        const acroFormRef = acroFormEntry.as(PdfObjectReference)?.resolve();
        if (!acroFormRef)
            return null;
        return acroFormRef.becomes(PdfAcroForm);
    }
    get pages() {
        const root = this.root;
        const pagesEntry = root.content.get('Pages');
        if (pagesEntry) {
            const pagesRef = pagesEntry.as(PdfObjectReference)?.resolve();
            if (pagesRef)
                return pagesRef.becomes(PdfPages);
        }
        const pages = new PdfPages();
        this.add(pages);
        root.content.set('Pages', pages.reference);
        return pages;
    }
    get header() {
        return this.revisions[0].header;
    }
    set header(comment) {
        if (comment)
            this.revisions[0].header = comment;
    }
    /**
     * Creates a PdfDocument from an array of PDF objects.
     * Parses objects into revisions based on EOF comments.
     *
     * @param objects - Array of PDF objects to construct the document from
     * @returns A new PdfDocument instance
     */
    static fromObjects(objects) {
        let header;
        const revisions = [];
        let currentObjects = [];
        for (const obj of objects) {
            if (obj instanceof PdfComment && obj.isVersionComment()) {
                header = obj;
                continue;
            }
            currentObjects.push(obj);
            if (obj instanceof PdfComment && obj.isEOFComment()) {
                revisions.push(new PdfRevision({ objects: currentObjects }));
                currentObjects = [];
            }
        }
        if (currentObjects.length > 0) {
            revisions.push(new PdfRevision({ objects: currentObjects }));
        }
        return new PdfDocument({ revisions, version: header });
    }
    /**
     * Starts a new revision for incremental updates.
     * Creates a new revision linked to the previous one.
     *
     * @returns The document instance for method chaining
     */
    startNewRevision() {
        const newRevision = new PdfRevision({ prev: this.latestRevision });
        this.revisions.push(newRevision);
        const lastStartXRef = this.objects.findLast((x) => x instanceof PdfStartXRef);
        if (lastStartXRef) {
            newRevision.xref.offset = lastStartXRef.offset.ref;
        }
        return this;
    }
    hasObjectInLatestRevision(obj) {
        return this.latestRevision.objects.includes(obj);
    }
    /**
     * Adds objects to the document's latest revision.
     * Automatically starts a new revision if the current one is locked.
     *
     * @param objects - PDF objects to add to the document
     */
    add(...objects) {
        if (this.latestRevision.locked) {
            this.startNewRevision();
        }
        for (const obj of objects) {
            this.wireResolvers(obj);
            if (this.hasObjectInLatestRevision(obj)) {
                continue;
            }
            this.latestRevision.addObject(obj);
        }
        // Auto-add any referenced-but-missing objects
        const missing = this.collectMissingReferences(...objects);
        if (missing.length > 0) {
            this.add(...missing);
        }
        this.update();
        for (const obj of objects) {
            obj.setModified(false);
        }
    }
    /**
     * Packs non-stream indirect objects into compressed ObjStm containers.
     * Objects that already have an object number are left unchanged.
     * Appearance streams and other PdfStream objects must NOT be passed here —
     * only non-stream indirect objects (dictionaries, arrays, etc.) are eligible.
     *
     * @param objects - Non-stream indirect objects to compress into ObjStm
     * @param batchSize - Maximum objects per ObjStm container (default 100)
     */
    addObjectsAsStream(objects, batchSize = 100) {
        if (objects.length === 0)
            return;
        for (let start = 0; start < objects.length; start += batchSize) {
            const batch = objects.slice(start, Math.min(start + batchSize, objects.length));
            // Pre-assign a contiguous block of object numbers for the batch
            const xref = this.latestRevision.xref;
            const startNum = xref.reserveObjectNumbers(batch.length);
            for (let i = 0; i < batch.length; i++) {
                if (!batch[i].inPdf()) {
                    batch[i].objectNumber = startNum + i;
                }
            }
            // Build the ObjStm and apply FlateDecode compression
            const objStream = PdfObjStream.fromObjects(batch);
            objStream.addFilter('FlateDecode');
            // Add the container — xref.addObject() auto-registers children as type-2 entries
            const container = new PdfIndirectObject({ content: objStream });
            this.add(container);
        }
    }
    /**
     * Gets the latest (most recent) revision of the document.
     *
     * @returns The latest PdfRevision
     * @throws Error if the revision for the last StartXRef cannot be found
     */
    get latestRevision() {
        const lastStartXRef = this.objects.findLast((x) => x instanceof PdfStartXRef);
        if (!lastStartXRef) {
            return this.revisions[this.revisions.length - 1];
        }
        const revision = this.revisions.find((rev) => rev.xref.offset === lastStartXRef.offset.ref) ??
            this.revisions.find((rev) => rev.xref.offset.equals(lastStartXRef.offset.value));
        if (!revision) {
            throw new Error('Cannot find revision for last StartXRef with offset ' +
                `${lastStartXRef.offset.value}. Options are: ` +
                this.revisions
                    .map((rev) => rev.xref.offset.value)
                    .join(', '));
        }
        return revision;
    }
    /**
     * Gets the cross-reference lookup table for the latest revision.
     *
     * @returns The PdfXrefLookup for the latest revision
     */
    get xrefLookup() {
        return this.latestRevision.xref;
    }
    /**
     * Gets the trailer dictionary from the cross-reference lookup.
     *
     * @returns The trailer dictionary containing document metadata references
     */
    get trailerDict() {
        return this.xrefLookup.trailerDict;
    }
    /**
     * Gets all objects across all revisions in the document.
     *
     * @returns A readonly array of all PDF objects
     */
    get objects() {
        return this.revisions.flatMap((rev) => rev.objects);
    }
    /**
     * Gets the encryption dictionary from the document if present.
     *
     * @returns The encryption dictionary object or undefined if not encrypted
     * @throws Error if the encryption dictionary reference points to a non-dictionary object
     */
    get encryptionDictionary() {
        const encryptionDictionaryRef = this.trailerDict
            .get('Encrypt')
            ?.as(PdfObjectReference);
        if (!encryptionDictionaryRef) {
            return undefined;
        }
        const encryptionDictObject = this.findUncompressedObject(encryptionDictionaryRef);
        if (!(encryptionDictObject?.content instanceof PdfDictionary)) {
            throw new Error(`Encryption dictionary object ${encryptionDictionaryRef.objectNumber} ${encryptionDictionaryRef.generationNumber} is not a dictionary, it is a ${encryptionDictObject?.content.objectType}`);
        }
        encryptionDictObject.encryptable = false;
        return encryptionDictObject;
    }
    get rootReference() {
        return this.root.reference;
    }
    /**
     * Gets the document catalog (root) dictionary, or creates one if it doesn't exist.
     *
     * @returns The root dictionary
     * @throws Error if the Root reference points to a non-dictionary object
     */
    get root() {
        const rootRef = this.trailerDict.get('Root')?.as(PdfObjectReference);
        if (!rootRef) {
            const rootObject = new PdfIndirectObject({
                content: new PdfDictionary(),
            });
            this.add(rootObject);
            this.trailerDict.set('Root', rootObject.reference);
            return rootObject;
        }
        const rootObject = this.findUncompressedObject(rootRef);
        if (!rootObject) {
            throw new Error('Root object not found');
        }
        if (!(rootObject?.content instanceof PdfDictionary)) {
            throw new Error(`Root object ${rootRef.objectNumber} ${rootRef.generationNumber} is not a dictionary, it is a ${rootObject?.content.objectType}`);
        }
        return rootObject;
    }
    /**
     * Gets the reference to the metadata stream from the document catalog.
     *
     * @returns The metadata stream reference or undefined if not present
     */
    get metadataStreamReference() {
        const root = this.root;
        if (!root) {
            return;
        }
        const metadataRef = root.content.get('Metadata')?.as(PdfObjectReference);
        if (!metadataRef) {
            return;
        }
        return metadataRef;
    }
    resetSecurityHandler() {
        this.securityHandler =
            this.originalSecurityHandler ?? this.getSecurityHandler();
    }
    getSecurityHandler() {
        const encryptionDictionaryRef = this.trailerDict
            .get('Encrypt')
            ?.as(PdfObjectReference);
        if (!encryptionDictionaryRef) {
            return undefined;
        }
        const encryptionDictObject = this.findUncompressedObject(encryptionDictionaryRef);
        if (!encryptionDictObject) {
            throw new Error('Encryption dictionary object not found');
        }
        if (!(encryptionDictObject?.content instanceof PdfDictionary)) {
            throw new Error(`Encryption dictionary object ${encryptionDictionaryRef.objectNumber} ${encryptionDictionaryRef.generationNumber} is not a dictionary, it is a ${encryptionDictObject?.content.objectType}`);
        }
        this.hasEncryptionDictionary = true;
        return createFromDictionary(encryptionDictObject.content, {
            documentId: this.trailerDict.get('ID'),
        });
    }
    initSecurityHandler(options) {
        if (this.securityHandler instanceof PdfStandardSecurityHandler) {
            const documentId = this.trailerDict.get('ID');
            options.password &&
                this.securityHandler.setPassword(options.password);
            options.ownerPassword &&
                this.securityHandler.setOwnerPassword(options.ownerPassword);
            documentId && this.securityHandler.setDocumentId(documentId);
            return;
        }
        this.securityHandler = new PdfV5SecurityHandler({
            password: options.password,
            ownerPassword: options.ownerPassword,
        });
    }
    /**
     * Sets the user password for document encryption.
     *
     * @param password - The user password to set
     * @throws Error if the security handler doesn't support password setting
     */
    setPassword(password) {
        if (this.securityHandler instanceof PdfStandardSecurityHandler) {
            this.securityHandler.setPassword(password);
        }
        else if (!this.securityHandler) {
            this.initSecurityHandler({ password });
        }
        else {
            throw new Error('Setting password is only supported for Standard Security Handler');
        }
    }
    /**
     * Sets the owner password for document encryption.
     *
     * @param ownerPassword - The owner password to set
     * @throws Error if the security handler doesn't support password setting
     */
    setOwnerPassword(ownerPassword) {
        if (this.securityHandler instanceof PdfStandardSecurityHandler) {
            this.securityHandler.setOwnerPassword(ownerPassword);
        }
        else if (!this.securityHandler) {
            this.initSecurityHandler({ ownerPassword });
        }
        else {
            throw new Error('Setting ownerPassword is only supported for Standard Security Handler');
        }
    }
    /**
     * Checks if a PDF object exists in the document.
     *
     * @param obj - The PDF object to check
     * @returns True if the object exists in the document
     */
    hasObject(obj) {
        return this.objects.includes(obj);
    }
    isObjectEncryptable(obj) {
        if (!this.securityHandler) {
            return false;
        }
        if (!obj.isEncryptable()) {
            return false;
        }
        if (obj.matchesReference(this.encryptionDictionary?.reference)) {
            return false;
        }
        if (!this.securityHandler.encryptMetadata &&
            obj.matchesReference(this.metadataStreamReference)) {
            return false;
        }
        return true;
    }
    commitIncrementalUpdates() {
        if (!this.incremental || this._committing) {
            return;
        }
        this._committing = true;
        try {
            for (const revision of this.revisions) {
                if (!revision.locked)
                    continue;
                for (const obj of revision.objects) {
                    if (!(obj instanceof PdfIndirectObject)) {
                        continue;
                    }
                    if (obj.isModified()) {
                        // Pre-register any new objects referenced by obj before
                        // cloning so the clone gets correct object numbers.
                        // Without this, proxy references (objectNumber = -1) get
                        // frozen into the clone as "-1 0 R" — a dangling reference.
                        const missing = this.collectMissingReferences(obj);
                        if (missing.length > 0) {
                            this.add(...missing);
                        }
                        const adding = obj.clone();
                        this.add(adding);
                        adding.setModified(false);
                        obj.setModified(false);
                    }
                }
            }
        }
        finally {
            this._committing = false;
        }
    }
    /**
     * Decrypts all encrypted object data in-place without removing
     * the encryption infrastructure. Useful in incremental mode where
     * the original (encrypted) bytes are preserved via cached tokens
     * but the live object data needs to be readable.
     */
    async decryptObjects() {
        if (!this.securityHandler) {
            return;
        }
        for (const object of this.objects) {
            if (!(object instanceof PdfIndirectObject)) {
                continue;
            }
            if (!this.isObjectEncryptable(object)) {
                continue;
            }
            await this.securityHandler.decryptObject(object);
        }
    }
    /**
     * Re-encrypts all objects and updates the document structure.
     * No-op if the document has no security handler (unencrypted document).
     */
    async finalize() {
        if (this._finalized) {
            throw new Error('Document has already been finalized');
        }
        this._finalized = true;
        this.update();
        if (this.securityHandler) {
            await this.encrypt();
        }
        this.update();
        await this.sign();
    }
    /**
     * Decrypts all encrypted objects in the document.
     * Removes the security handler and encryption dictionary after decryption.
     */
    async decrypt() {
        if (!this.securityHandler) {
            return;
        }
        await this.decryptObjects();
        this.hasEncryptionDictionary = false;
        this.securityHandler = undefined;
        const encryptionDict = this.encryptionDictionary;
        this.trailerDict.delete('Encrypt');
        if (encryptionDict) {
            this.deleteObject(encryptionDict);
        }
        this.update();
    }
    /**
     * Encrypts all encryptable objects using the security handler.
     * Re-uses the existing encryption dictionary or creates one if needed,
     * propagating it to all revisions.
     */
    async encrypt() {
        if (!this.securityHandler) {
            return;
        }
        const addingEncryption = !this.hasEncryptionDictionary;
        const wasIncremental = this.incremental;
        if (addingEncryption) {
            this.setIncremental(false);
        }
        this.initSecurityHandler({});
        await this.securityHandler.write();
        for (const object of this.objects) {
            if (!(object instanceof PdfIndirectObject)) {
                continue;
            }
            if (!this.isObjectEncryptable(object)) {
                continue;
            }
            await this.securityHandler.encryptObject(object);
            object.setModified(false);
        }
        if (addingEncryption) {
            const encryptionDictObject = new PdfIndirectObject({
                content: this.securityHandler.dict,
                encryptable: false,
            });
            this.add(encryptionDictObject);
            this.hasEncryptionDictionary = true;
            for (const revision of this.revisions) {
                revision.xref.trailerDict.set('Encrypt', encryptionDictObject.reference);
                if (!revision.xref.trailerDict.get('ID')) {
                    revision.xref.trailerDict.set('ID', this.securityHandler.getDocumentId());
                }
            }
            this.setIncremental(wasIncremental);
        }
        this.update();
    }
    /**
     * Finds a compressed object by its object number within an object stream.
     *
     * @param options - Object identifier with objectNumber and optional generationNumber
     * @returns The found indirect object or undefined if not found
     * @throws Error if the object cannot be found in the expected object stream
     */
    findCompressedObject(options) {
        const xrefEntry = this.xrefLookup.getObject(options.objectNumber);
        if (!(xrefEntry instanceof PdfXRefStreamCompressedEntry)) {
            throw new Error('Cannot find object inside object stream via PdfDocument.findObject');
        }
        const streamNumber = xrefEntry.objectStreamNumber.value;
        let objectStream = this._objStreamCache.get(streamNumber);
        if (!objectStream) {
            const objectStreamIndirect = this.findUncompressedObject({
                objectNumber: streamNumber,
            });
            if (!objectStreamIndirect) {
                throw new Error(`Cannot find object stream ${streamNumber} for object ${options.objectNumber}`);
            }
            objectStream = objectStreamIndirect.content
                .as(PdfStream)
                .parseAs(PdfObjStream);
            this._objStreamCache.set(streamNumber, objectStream);
        }
        return objectStream.getObject({ objectNumber: options.objectNumber });
    }
    /**
     * Finds an uncompressed indirect object by its object number.
     *
     * @param options - Object identifier with objectNumber and optional generationNumber
     * @returns The found indirect object or undefined if not found
     * @throws FoundCompressedObjectError if the object is compressed (in an object stream)
     */
    findUncompressedObject(options) {
        const xrefEntry = this.xrefLookup.getObject(options.objectNumber);
        if (xrefEntry instanceof PdfXRefStreamCompressedEntry) {
            throw new FoundCompressedObjectError(`Cannot find object ${options.objectNumber} inside object stream via PdfDocument.findObject`);
        }
        if (!xrefEntry ||
            (options.generationNumber !== undefined &&
                xrefEntry.generationNumber.value !== options.generationNumber)) {
            return undefined;
        }
        const found = this.objects.find((obj) => obj instanceof PdfIndirectObject &&
            obj.objectNumber === options.objectNumber &&
            (options.generationNumber === undefined ||
                obj.generationNumber === options.generationNumber) &&
            obj.offset.equals(xrefEntry.byteOffset.ref));
        return found;
    }
    /**
     * Reads and optionally decrypts an object by its object number.
     * Handles both compressed and uncompressed objects.
     *
     * @param options - Object lookup options
     * @param options.objectNumber - The object number to find
     * @param options.generationNumber - Optional generation number filter
     * @param options.allowUnindexed - If true, searches unindexed objects as fallback
     * @returns A cloned and decrypted copy of the object, or undefined if not found
     */
    readObject(options) {
        let foundObject;
        try {
            foundObject = this.findUncompressedObject(options);
        }
        catch (e) {
            if (e instanceof FoundCompressedObjectError) {
                foundObject = this.findCompressedObject(options);
            }
            else {
                throw e;
            }
        }
        if (!foundObject && options.allowUnindexed) {
            foundObject = this.objects.find((obj) => obj instanceof PdfIndirectObject &&
                obj.objectNumber === options.objectNumber &&
                (options.generationNumber === undefined ||
                    obj.generationNumber === options.generationNumber));
        }
        if (!foundObject) {
            return undefined;
        }
        if (options.cloned) {
            foundObject = foundObject.clone();
        }
        return foundObject;
    }
    /**
     * Finds the revision that contains a given PDF object.
     * Useful for determining the origin of an object across multiple revisions.
     * @param obj - The PDF object to find the revision for
     * @returns The PdfRevision that contains the object, or undefined if not found in any revision
     */
    findRevisionForObject(obj) {
        return this.revisions.find((rev) => rev.objects.includes(obj));
    }
    /**
     * Deletes an object from all revisions in the document.
     *
     * @param obj - The PDF object to delete
     */
    deleteObject(obj) {
        if (!obj)
            return;
        for (const revision of this.revisions) {
            revision.deleteObject(obj);
        }
        this.update();
    }
    /**
     * Sets the PDF version for the document.
     *
     * @param version - The PDF version string (e.g., '1.7', '2.0')
     * @throws Error if attempting to change version after objects have been added in incremental mode
     */
    setVersion(version) {
        if (this.revisions[0].locked) {
            throw new Error('Cannot change PDF version in incremental mode after objects have been added');
        }
        this.header = PdfComment.versionComment(version);
    }
    /**
     * Sets whether the document should use incremental updates.
     * When true, locks all existing revisions to preserve original content.
     *
     * @param value - True to enable incremental mode, false to disable. Defaults to true.
     */
    setIncremental(value = true) {
        if (value === this.isIncremental()) {
            return;
        }
        this.incremental = value;
        for (const revision of this.revisions) {
            revision.locked = value;
        }
    }
    /**
     * Checks if the document is in incremental mode.
     *
     * @returns True if all revisions are locked for incremental updates
     */
    isIncremental() {
        return this.incremental;
    }
    /**
     * Returns tokens paired with their source objects.
     * Useful for debugging and analysis of document structure.
     *
     * @returns Array of token-object pairs
     */
    tokensWithObjects() {
        const documentTokens = this.objects.flatMap((obj) => {
            const tokens = obj.toTokens();
            if (tokens.length > 0 &&
                !(tokens[tokens.length - 1] instanceof PdfWhitespaceToken)) {
                tokens.push(PdfWhitespaceToken.NEWLINE);
            }
            return tokens.map((token) => ({ token, object: obj }));
        });
        return documentTokens;
    }
    tokenize() {
        return this.tokensWithObjects().map(({ token }) => token);
    }
    wireResolvers(...objects) {
        const seen = new Set();
        const walk = (obj) => {
            if (seen.has(obj))
                return;
            seen.add(obj);
            if (obj instanceof PdfObjectReference &&
                !(obj instanceof PdfIndirectObject)) {
                obj.resolver = this;
            }
            else if (obj instanceof PdfStream) {
                walk(obj.header);
            }
            else if (obj instanceof PdfDictionary) {
                for (const [, value] of obj.entries()) {
                    if (value)
                        walk(value);
                }
            }
            else if (obj instanceof PdfArray) {
                for (const item of obj.items) {
                    walk(item);
                }
            }
            else if (obj instanceof PdfIndirectObject) {
                walk(obj.content);
            }
        };
        for (const obj of objects) {
            walk(obj);
        }
    }
    collectMissingReferences(...objects) {
        const seen = new Set();
        const missing = [];
        const queue = [...objects];
        while (queue.length > 0) {
            const obj = queue.pop();
            if (seen.has(obj))
                continue;
            seen.add(obj);
            if (obj instanceof PdfObjectReference &&
                !(obj instanceof PdfIndirectObject)) {
                if (obj.resolver) {
                    try {
                        const resolved = obj.resolve();
                        if (resolved instanceof PdfIndirectObject &&
                            !resolved.inPdf()) {
                            missing.push(resolved);
                            queue.push(resolved);
                        }
                    }
                    catch {
                        // Skip unresolvable references
                    }
                }
            }
            else if (obj instanceof PdfStream) {
                queue.push(obj.header);
            }
            else if (obj instanceof PdfDictionary) {
                for (const [, value] of obj.entries()) {
                    if (value)
                        queue.push(value);
                }
            }
            else if (obj instanceof PdfArray) {
                for (const item of obj.items)
                    queue.push(item);
            }
            else if (obj instanceof PdfIndirectObject) {
                queue.push(obj.content);
            }
        }
        return missing;
    }
    linkRevisions() {
        const xrefLookups = this.revisions.map((rev) => rev.xref);
        const indirectObjects = this.objects.filter((x) => x instanceof PdfIndirectObject);
        for (const revision of this.revisions) {
            revision.xref.linkPrev(xrefLookups);
            // Walk the entire prev chain to link all xref entries
            let xref = revision.xref;
            while (xref) {
                xref.linkIndirectObjects(indirectObjects);
                xref = xref.prev;
            }
        }
    }
    linkOffsets() {
        const refMap = new Map();
        const tokens = this.toTokens();
        for (let i = 0; i < tokens.length; i++) {
            const token = tokens[i];
            let main;
            let other;
            if (token instanceof PdfByteOffsetToken) {
                main = token.value;
            }
            else if (token instanceof PdfXRefTableEntryToken) {
                other = token.offset.ref;
            }
            else if (token instanceof PdfNumberToken && token.isByteToken) {
                other = token.ref;
            }
            if (!other && !main) {
                continue;
            }
            const id = (main ?? other).resolve();
            if (!refMap.has(id)) {
                refMap.set(id, { main: main, others: new Set() });
            }
            if (main)
                refMap.get(id).main = main;
            if (other)
                refMap.get(id).others.add(other);
        }
        for (const [, { main, others }] of refMap) {
            if (!main)
                continue;
            for (const other of others ?? []) {
                other.update(main);
            }
        }
    }
    calculateOffsets() {
        const serializer = new PdfTokenSerializer();
        serializer.feedMany(this.toTokens());
        serializer.calculateOffsets();
        this.linkOffsets();
    }
    updateRevisions() {
        let modified = false;
        this.revisions.forEach((rev) => {
            if (rev.isModified()) {
                modified = true;
            }
            if (modified) {
                rev.update();
            }
        });
        // Always rebuild xref binary data to reflect recalculated offsets
        // This walks the entire prev chain for each revision
        for (const rev of this.revisions) {
            let xref = rev.xref;
            while (xref) {
                xref.update();
                xref = xref.prev;
            }
        }
    }
    /**
     * Performs a full update cycle to ensure all revisions are consistent and offsets are correct.
     */
    update() {
        if (this._updating)
            return;
        this._updating = true;
        try {
            this.commitIncrementalUpdates();
            this.flushResolvedCache();
            this.registerNewReferences();
            this.calculateOffsets();
            this.updateRevisions();
            // Second pass: xref binary may have changed size (e.g. FlateDecode removed
            // from xref stream), shifting objects that follow it. Recalculate so entry
            // byteOffset refs hold the new positions, then rebuild the xref binary once
            // more so the baked bytes match those positions.
            this.calculateOffsets();
            this.updateRevisions();
            // Third pass: confirm positions are stable (xref binary size should not
            // change again because W widths and entry count are the same).
            this.calculateOffsets();
        }
        finally {
            this._updating = false;
        }
    }
    /**
     * Walks all objects in the document and registers any newly created
     * PdfIndirectObjects that are referenced but not yet part of the document
     * (e.g. appearance streams created by generateAppearance).
     */
    registerNewReferences() {
        const missing = this.collectMissingReferences(...this.latestRevision.objects);
        if (missing.length > 0) {
            this.add(...missing);
        }
    }
    flushResolvedCache() {
        const existing = new Set(this.objects);
        for (const [, obj] of this._resolvedCache) {
            if (obj.isModified() && !existing.has(obj)) {
                this.add(obj);
                existing.add(obj);
            }
        }
    }
    async sign() {
        if (this._signed) {
            throw new Error('Document has already been signed');
        }
        if (!this.signer) {
            return;
        }
        await this.signer.sign();
        this._signed = true;
    }
    /**
     * Serializes the document to a byte array.
     *
     * @returns The PDF document as a Uint8Array
     */
    toBytes() {
        this.update();
        return concatUint8Arrays(this.revisions.map((x) => x.toBytes()));
    }
    /**
     * Creates a deep copy of the document.
     *
     * @returns A cloned PdfDocument instance
     */
    cloneImpl() {
        const clonedRevisions = this.revisions.map((rev) => rev.clone());
        return new PdfDocument({
            revisions: clonedRevisions,
            version: this.header?.clone(),
            securityHandler: this.securityHandler,
        });
    }
    toJSON() {
        return {
            type: 'document',
            revisions: this.revisions.map((rev) => rev.toJSON()),
        };
    }
    /**
     * Creates a PdfDocument from a byte stream.
     *
     * @param input - Async or sync iterable of byte arrays
     * @returns A promise that resolves to the parsed PdfDocument
     */
    static fromBytes(input, options) {
        return PdfReader.fromBytes(input, options);
    }
    isModified() {
        return (super.isModified() || this.revisions.some((rev) => rev.isModified()));
    }
    /**
     * Verifies all digital signatures in the document.
     *
     * @returns A promise that resolves to the verification result
     */
    async verifySignatures() {
        return await this.signer.verify();
    }
}
