// pki-lite/dist/pkcs7/DigestedData.js
import { asn1js, PkiBase } from '../core/PkiBase.js';
/**
 * Represents a PKCS#7 DigestedData structure.
 *
 * @asn
 * ```asn
 * DigestedData ::= SEQUENCE {
 *      version INTEGER,
 *      digestAlgorithm DigestAlgorithmIdentifier,
 *      encapContentInfo EncapsulatedContentInfo,
 *      digest Digest
 * }
 * ```
 */
export class DigestedData extends PkiBase {
    version;
    digestAlgorithm;
    encapContentInfo;
    digest;
    constructor(options) {
        super();
        const { version, digestAlgorithm, encapContentInfo, digest } = options;
        this.version = version;
        this.digestAlgorithm = digestAlgorithm;
        this.encapContentInfo = encapContentInfo;
        this.digest = digest;
    }
    /**
     * Converts the DigestedData to an ASN.1 structure.
     */
    toAsn1() {
        return new asn1js.Sequence({
            value: [
                new asn1js.Integer({ value: this.version }),
                this.digestAlgorithm.toAsn1(),
                this.encapContentInfo.toAsn1(),
                new asn1js.OctetString({ valueHex: this.digest }),
            ],
        });
    }
}
