// pki-lite/dist/pkcs7/IssuerAndSerialNumber.js
import { asn1js, PkiBase } from '../core/PkiBase.js';
import { RDNSequence } from '../x509/RDNSequence.js';
import { Integer } from '../asn1/Integer.js';
import { Certificate } from '../x509/Certificate.js';
import { Asn1ParseError } from '../core/errors/Asn1ParseError.js';
/**
 * Represents an IssuerAndSerialNumber structure used in PKCS#7.
 *
 * @asn
 * ```asn
 * IssuerAndSerialNumber ::= SEQUENCE {
 *      issuer Name,
 *      serialNumber CertificateSerialNumber
 * }
 *
 * CertificateSerialNumber ::= INTEGER
 * ```
 */
export class IssuerAndSerialNumber extends PkiBase {
    issuer;
    serialNumber;
    constructor(options) {
        super();
        const { issuer, serialNumber } = options;
        this.issuer = issuer;
        this.serialNumber = new Integer({ value: serialNumber });
    }
    /**
     * Converts the IssuerAndSerialNumber to an ASN.1 structure.
     */
    toAsn1() {
        return new asn1js.Sequence({
            value: [this.issuer.toAsn1(), this.serialNumber.toAsn1()],
        });
    }
    /**
     * Creates an IssuerAndSerialNumber from an ASN.1 structure.
     *
     * @param asn1 The ASN.1 structure
     * @returns An IssuerAndSerialNumber
     */
    static fromAsn1(asn1) {
        if (!(asn1 instanceof asn1js.Sequence)) {
            throw new Asn1ParseError('Invalid ASN.1 structure: expected SEQUENCE but got ' +
                asn1.constructor.name);
        }
        if (asn1.valueBlock.value.length !== 2) {
            throw new Asn1ParseError('Invalid ASN.1 structure: expected 2 elements');
        }
        const issuerAsn1 = asn1.valueBlock.value[0];
        const serialNumberAsn1 = asn1.valueBlock.value[1];
        if (!(issuerAsn1 instanceof asn1js.Sequence)) {
            throw new Asn1ParseError('Invalid ASN.1 structure: issuer must be a SEQUENCE');
        }
        const issuer = RDNSequence.fromAsn1(issuerAsn1);
        const serialNumber = Integer.fromAsn1(serialNumberAsn1);
        return new IssuerAndSerialNumber({ issuer, serialNumber });
    }
    matchesCertificate(cert) {
        if (cert instanceof Certificate) {
            return (cert.tbsCertificate.issuer.equals(this.issuer) &&
                cert.tbsCertificate.serialNumber.equals(this.serialNumber));
        }
        else {
            throw new Error('Unsupported certificate choice for matching');
        }
    }
    toHumanString() {
        return `${this.issuer.toHumanString()}, SN=${this.serialNumber.toHexString()}`;
    }
}
